{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Hide FutureWarning\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import pyranges as pr\n",
    "\n",
    "path = '/project/hfa_work/ENCODE/data'\n",
    "alignment_path = 'alignment_v45'\n",
    "genome_file = 'GRCh38.p14.genome.fa'\n",
    "genome_path = os.path.join(path, 'gencode_human/version_45', genome_file)\n",
    "window_padding = 100"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_file = 'reads/metadata_tissue.tsv'\n",
    "metadata = pd.read_csv(os.path.join(path, metadata_file), sep='\\t')\n",
    "tissues = metadata['group'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['aorta', 'brain', 'colon', 'heart', 'lung', 'muscle']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CAGE_path = '/project/hfa_work/ENCODE/code/snakemake-pipeline/resources/CAGE/'\n",
    "# filter out tissues with no corresponding CAGE data\n",
    "tissues = [tissue for tissue in tissues if os.path.exists(os.path.join(CAGE_path, f'{tissue}.bed'))]\n",
    "tissues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tool</th>\n",
       "      <th>tissue</th>\n",
       "      <th>gtf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>flair</td>\n",
       "      <td>aorta</td>\n",
       "      <td>/project/hfa_work/ENCODE/code/snakemake-pipeli...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>flair</td>\n",
       "      <td>brain</td>\n",
       "      <td>/project/hfa_work/ENCODE/code/snakemake-pipeli...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>flair</td>\n",
       "      <td>colon</td>\n",
       "      <td>/project/hfa_work/ENCODE/code/snakemake-pipeli...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>flair</td>\n",
       "      <td>heart</td>\n",
       "      <td>/project/hfa_work/ENCODE/code/snakemake-pipeli...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>flair</td>\n",
       "      <td>lung</td>\n",
       "      <td>/project/hfa_work/ENCODE/code/snakemake-pipeli...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>flair</td>\n",
       "      <td>muscle</td>\n",
       "      <td>/project/hfa_work/ENCODE/code/snakemake-pipeli...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>isotools</td>\n",
       "      <td>aorta</td>\n",
       "      <td>/project/hfa_work/ENCODE/code/snakemake-pipeli...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>isotools</td>\n",
       "      <td>brain</td>\n",
       "      <td>/project/hfa_work/ENCODE/code/snakemake-pipeli...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>isotools</td>\n",
       "      <td>colon</td>\n",
       "      <td>/project/hfa_work/ENCODE/code/snakemake-pipeli...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>isotools</td>\n",
       "      <td>heart</td>\n",
       "      <td>/project/hfa_work/ENCODE/code/snakemake-pipeli...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>isotools</td>\n",
       "      <td>lung</td>\n",
       "      <td>/project/hfa_work/ENCODE/code/snakemake-pipeli...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>isotools</td>\n",
       "      <td>muscle</td>\n",
       "      <td>/project/hfa_work/ENCODE/code/snakemake-pipeli...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        tool  tissue                                                gtf\n",
       "0      flair   aorta  /project/hfa_work/ENCODE/code/snakemake-pipeli...\n",
       "1      flair   brain  /project/hfa_work/ENCODE/code/snakemake-pipeli...\n",
       "2      flair   colon  /project/hfa_work/ENCODE/code/snakemake-pipeli...\n",
       "3      flair   heart  /project/hfa_work/ENCODE/code/snakemake-pipeli...\n",
       "4      flair    lung  /project/hfa_work/ENCODE/code/snakemake-pipeli...\n",
       "5      flair  muscle  /project/hfa_work/ENCODE/code/snakemake-pipeli...\n",
       "6   isotools   aorta  /project/hfa_work/ENCODE/code/snakemake-pipeli...\n",
       "7   isotools   brain  /project/hfa_work/ENCODE/code/snakemake-pipeli...\n",
       "8   isotools   colon  /project/hfa_work/ENCODE/code/snakemake-pipeli...\n",
       "9   isotools   heart  /project/hfa_work/ENCODE/code/snakemake-pipeli...\n",
       "10  isotools    lung  /project/hfa_work/ENCODE/code/snakemake-pipeli...\n",
       "11  isotools  muscle  /project/hfa_work/ENCODE/code/snakemake-pipeli..."
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# isotools_path = '/project/hfa_work/ENCODE/code/notebooks/results/isotools_'\n",
    "fofn_path = '/project/hfa_work/ENCODE/code/snakemake-pipeline/results/tissues_gtf.fofn'\n",
    "# fofn_path = '/project/hfa_work/ENCODE/code/snakemake-pipeline/results/flair/tissue_gtfs.fofn'\n",
    "\n",
    "tool_gtfs = pd.read_csv(fofn_path, sep='\\t', header=None, names=['tool', 'tissue', 'gtf'])\n",
    "# check that there are no duplicates of tool + tissue\n",
    "assert tool_gtfs.groupby(['tool', 'tissue']).size().max() == 1\n",
    "# filter for tissues with CAGE data\n",
    "tool_gtfs = tool_gtfs[tool_gtfs['tissue'].isin(tissues)]\n",
    "tool_gtfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "display_tool = 'flair'\n",
    "display_tissue = 'brain'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "CAGE_data = {}\n",
    "for tissue in tissues:\n",
    "    CAGE_data[tissue] = pr.read_bed(os.path.join(CAGE_path, f'{tissue}.bed'))\n",
    "    CAGE_data[tissue] = CAGE_data[tissue].assign('CAGE_score', lambda x: x['Score']).drop(drop='Score')\n",
    "\n",
    "tool_gtfs['pr'] = tool_gtfs['gtf'].apply(lambda x: pr.read_gtf(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>transcript</td>\n",
       "      <td>175387</td>\n",
       "      <td>175644</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:175000</td>\n",
       "      <td>m64109_220203_225821/159646893/ccs</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>175387</td>\n",
       "      <td>175644</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:175000</td>\n",
       "      <td>m64109_220203_225821/159646893/ccs</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>transcript</td>\n",
       "      <td>197700</td>\n",
       "      <td>199762</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:197000</td>\n",
       "      <td>m64109_220205_045735/17237446/ccs</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>197700</td>\n",
       "      <td>199762</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:197000</td>\n",
       "      <td>m64109_220205_045735/17237446/ccs</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>transcript</td>\n",
       "      <td>265404</td>\n",
       "      <td>267777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:265000</td>\n",
       "      <td>m64109_220205_045735/39125266/ccs</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>821086</th>\n",
       "      <td>chrY</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>307731</td>\n",
       "      <td>307880</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292358.1</td>\n",
       "      <td>m64109_220203_225821/137103082/ccs</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>821087</th>\n",
       "      <td>chrY</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>311418</td>\n",
       "      <td>311627</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292358.1</td>\n",
       "      <td>m64109_220203_225821/137103082/ccs</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>821088</th>\n",
       "      <td>chrY</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>312765</td>\n",
       "      <td>315020</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292358.1</td>\n",
       "      <td>m64109_220203_225821/137103082/ccs</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>821089</th>\n",
       "      <td>chrY</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>316913</td>\n",
       "      <td>317051</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292358.1</td>\n",
       "      <td>m64109_220203_225821/137103082/ccs</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>821090</th>\n",
       "      <td>chrY</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>318438</td>\n",
       "      <td>318796</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292358.1</td>\n",
       "      <td>m64109_220203_225821/137103082/ccs</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>821091 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +6    |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | FLAIR      | transcript   | 175387    | 175644    | ...   |\n",
       "| chr1         | FLAIR      | exon         | 175387    | 175644    | ...   |\n",
       "| chr1         | FLAIR      | transcript   | 197700    | 199762    | ...   |\n",
       "| chr1         | FLAIR      | exon         | 197700    | 199762    | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chrY         | FLAIR      | exon         | 311418    | 311627    | ...   |\n",
       "| chrY         | FLAIR      | exon         | 312765    | 315020    | ...   |\n",
       "| chrY         | FLAIR      | exon         | 316913    | 317051    | ...   |\n",
       "| chrY         | FLAIR      | exon         | 318438    | 318796    | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 821,091 rows and 11 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "6 hidden columns: Score, Strand, Frame, gene_id, transcript_id, exon_number"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tool_gtfs.loc[(tool_gtfs['tool'] == display_tool) & (tool_gtfs['tissue'] == display_tissue), 'pr'].values[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Find TSS based on strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter out all positions where extending the window would go out of bounds\n",
    "tool_gtfs['pr'] = tool_gtfs['pr'].apply(lambda x: x.subset(lambda y: y.Start - window_padding >= 0)) # and y.end + window_padding <= x.chromsizes[0]))\n",
    "# no chromsizes yet\n",
    "# tool_gtfs['pr'] = tool_gtfs['pr'].apply(lambda x: x.filter(lambda y: y.start - window_padding >= 0 and y.end + window_padding <= x.chromsizes[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "flair aorta 19826 duplicate positions dropped\n",
      "flair brain 32504 duplicate positions dropped\n",
      "flair colon 21453 duplicate positions dropped\n",
      "flair heart 95768 duplicate positions dropped\n",
      "flair lung 50988 duplicate positions dropped\n",
      "flair muscle 17123 duplicate positions dropped\n",
      "isotools aorta 3961 duplicate positions dropped\n",
      "isotools brain 7747 duplicate positions dropped\n",
      "isotools colon 3814 duplicate positions dropped\n",
      "isotools heart 32430 duplicate positions dropped\n",
      "isotools lung 12135 duplicate positions dropped\n",
      "isotools muscle 3470 duplicate positions dropped\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>175387</td>\n",
       "      <td>175388</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>197700</td>\n",
       "      <td>197701</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>265404</td>\n",
       "      <td>265405</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>266854</td>\n",
       "      <td>266855</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>271731</td>\n",
       "      <td>271732</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111403</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1392112</td>\n",
       "      <td>1392113</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111404</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1452928</td>\n",
       "      <td>1452929</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111405</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2500975</td>\n",
       "      <td>2500976</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111406</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2500649</td>\n",
       "      <td>2500650</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111407</th>\n",
       "      <td>chrY</td>\n",
       "      <td>318795</td>\n",
       "      <td>318796</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>111408 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+\n",
       "| Chromosome   | Start     | End       | Strand       |\n",
       "| (category)   | (int64)   | (int64)   | (category)   |\n",
       "|--------------+-----------+-----------+--------------|\n",
       "| chr1         | 175387    | 175388    | +            |\n",
       "| chr1         | 197700    | 197701    | +            |\n",
       "| chr1         | 265404    | 265405    | +            |\n",
       "| chr1         | 266854    | 266855    | +            |\n",
       "| ...          | ...       | ...       | ...          |\n",
       "| chrY         | 1452928   | 1452929   | -            |\n",
       "| chrY         | 2500975   | 2500976   | -            |\n",
       "| chrY         | 2500649   | 2500650   | -            |\n",
       "| chrY         | 318795    | 318796    | -            |\n",
       "+--------------+-----------+-----------+--------------+\n",
       "Stranded PyRanges object has 111,408 rows and 4 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def tss_from_transcriptome(tool, tissue, transcriptome):\n",
    "    transcriptome = transcriptome[transcriptome.Feature == 'transcript']\n",
    "    tss_sites = transcriptome.five_end()\n",
    "    # plus = transcriptome[transcriptome.Strand == '+']\n",
    "    # plus.End = plus.Start + 1\n",
    "    # minus = transcriptome[transcriptome.Strand == '-']\n",
    "    # minus.Start = minus.End - 1\n",
    "    tss_sites_deduplicated = tss_sites.drop_duplicate_positions()\n",
    "    tss_beds(tool, tissue, tss_sites_deduplicated)\n",
    "    print(tool, tissue, tss_sites.df.shape[0] - tss_sites_deduplicated.df.shape[0], \"duplicate positions dropped\")\n",
    "    return tss_sites_deduplicated.drop()\n",
    "\n",
    "def tss_beds(tool, tissue, tss_sites_deduplicated):\n",
    "    # two beds per transcriptome, one for window inside of transcript and one outside\n",
    "    # TODO: Does this need a check for the end too? Would require a chromsizes file\n",
    "    outside = tss_sites_deduplicated.extend({\"5\": window_padding})\n",
    "    outside.to_bed(f'/project/hfa_work/ENCODE/code/snakemake-pipeline/results/tss_ratio/tss_outside_{tool}_{tissue}.bed', keep=False)\n",
    "    inside = tss_sites_deduplicated.extend({\"3\": window_padding})\n",
    "    inside.to_bed(f'/project/hfa_work/ENCODE/code/snakemake-pipeline/results/tss_ratio/tss_inside_{tool}_{tissue}.bed', keep=False)\n",
    "\n",
    "tool_gtfs['tss_sites'] = tool_gtfs.apply(lambda x: tss_from_transcriptome(x['tool'], x['tissue'], x['pr']), axis=1)\n",
    "tool_gtfs.loc[(tool_gtfs['tool'] == display_tool) & (tool_gtfs['tissue'] == display_tissue), 'tss_sites'].values[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CAGE support"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Strand</th>\n",
       "      <th>CAGE_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629179</td>\n",
       "      <td>629180</td>\n",
       "      <td>chr1:629179..629180,+</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629210</td>\n",
       "      <td>629211</td>\n",
       "      <td>chr1:629210..629211,+</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629212</td>\n",
       "      <td>629213</td>\n",
       "      <td>chr1:629212..629213,+</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629620</td>\n",
       "      <td>629621</td>\n",
       "      <td>chr1:629620..629621,+</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629625</td>\n",
       "      <td>629626</td>\n",
       "      <td>chr1:629625..629626,+</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3059674</th>\n",
       "      <td>chrY</td>\n",
       "      <td>20782681</td>\n",
       "      <td>20782682</td>\n",
       "      <td>chrY:20782681..20782682,-</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3059675</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21303006</td>\n",
       "      <td>21303007</td>\n",
       "      <td>chrY:21303006..21303007,-</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3059676</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25626516</td>\n",
       "      <td>25626517</td>\n",
       "      <td>chrY:25626516..25626517,-</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3059677</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26401653</td>\n",
       "      <td>26401654</td>\n",
       "      <td>chrY:26401653..26401654,-</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3059678</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26493515</td>\n",
       "      <td>26493516</td>\n",
       "      <td>chrY:26493515..26493516,-</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3059679 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+---------------------------+-------+\n",
       "| Chromosome   | Start     | End       | Name                      | +2    |\n",
       "| (category)   | (int64)   | (int64)   | (object)                  | ...   |\n",
       "|--------------+-----------+-----------+---------------------------+-------|\n",
       "| chr1         | 629179    | 629180    | chr1:629179..629180,+     | ...   |\n",
       "| chr1         | 629210    | 629211    | chr1:629210..629211,+     | ...   |\n",
       "| chr1         | 629212    | 629213    | chr1:629212..629213,+     | ...   |\n",
       "| chr1         | 629620    | 629621    | chr1:629620..629621,+     | ...   |\n",
       "| ...          | ...       | ...       | ...                       | ...   |\n",
       "| chrY         | 21303006  | 21303007  | chrY:21303006..21303007,- | ...   |\n",
       "| chrY         | 25626516  | 25626517  | chrY:25626516..25626517,- | ...   |\n",
       "| chrY         | 26401653  | 26401654  | chrY:26401653..26401654,- | ...   |\n",
       "| chrY         | 26493515  | 26493516  | chrY:26493515..26493516,- | ...   |\n",
       "+--------------+-----------+-----------+---------------------------+-------+\n",
       "Stranded PyRanges object has 3,059,679 rows and 6 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "2 hidden columns: Strand, CAGE_score"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CAGE_data[display_tissue]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>CAGE_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629079</td>\n",
       "      <td>629280</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629110</td>\n",
       "      <td>629311</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629112</td>\n",
       "      <td>629313</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629520</td>\n",
       "      <td>629721</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629525</td>\n",
       "      <td>629726</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3059674</th>\n",
       "      <td>chrY</td>\n",
       "      <td>20782581</td>\n",
       "      <td>20782782</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3059675</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21302906</td>\n",
       "      <td>21303107</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3059676</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25626416</td>\n",
       "      <td>25626617</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3059677</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26401553</td>\n",
       "      <td>26401754</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3059678</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26493415</td>\n",
       "      <td>26493616</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3059679 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+--------------+\n",
       "| Chromosome   | Start     | End       | Strand       | CAGE_score   |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | (int64)      |\n",
       "|--------------+-----------+-----------+--------------+--------------|\n",
       "| chr1         | 629079    | 629280    | +            | 1            |\n",
       "| chr1         | 629110    | 629311    | +            | 1            |\n",
       "| chr1         | 629112    | 629313    | +            | 1            |\n",
       "| chr1         | 629520    | 629721    | +            | 1            |\n",
       "| ...          | ...       | ...       | ...          | ...          |\n",
       "| chrY         | 21302906  | 21303107  | -            | 1            |\n",
       "| chrY         | 25626416  | 25626617  | -            | 1            |\n",
       "| chrY         | 26401553  | 26401754  | -            | 1            |\n",
       "| chrY         | 26493415  | 26493616  | -            | 1            |\n",
       "+--------------+-----------+-----------+--------------+--------------+\n",
       "Stranded PyRanges object has 3,059,679 rows and 5 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extend positions by 20bp in each direction\n",
    "CAGE_windows = {tissue: CAGE_data[tissue][['CAGE_score']].extend(window_padding) for tissue in tissues}\n",
    "CAGE_windows[display_tissue]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_CAGE_Support(tss_sites, CAGE_windows):\n",
    "    merged = tss_sites.join(CAGE_windows, strandedness='same', how='left')\n",
    "    clustered = merged.cluster(slack=-1).boundaries(\"Cluster\", agg={\"CAGE_score\": \"sum\"}).drop(drop=[\"Cluster\"])\n",
    "    return clustered\n",
    "\n",
    "tool_gtfs['CAGE_clusters'] = tool_gtfs.apply(lambda x: add_CAGE_Support(x['tss_sites'], CAGE_windows[x['tissue']]), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>CAGE_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>175387</td>\n",
       "      <td>175388</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>197700</td>\n",
       "      <td>197701</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>265404</td>\n",
       "      <td>265405</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>266854</td>\n",
       "      <td>266855</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>271731</td>\n",
       "      <td>271732</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111403</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1452928</td>\n",
       "      <td>1452929</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111404</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1596449</td>\n",
       "      <td>1596450</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111405</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2500649</td>\n",
       "      <td>2500650</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111406</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2500975</td>\n",
       "      <td>2500976</td>\n",
       "      <td>-</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111407</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57201427</td>\n",
       "      <td>57201428</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>111408 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+--------------+\n",
       "| Chromosome   | Start     | End       | Strand       | CAGE_score   |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | (int64)      |\n",
       "|--------------+-----------+-----------+--------------+--------------|\n",
       "| chr1         | 175387    | 175388    | +            | -1           |\n",
       "| chr1         | 197700    | 197701    | +            | -1           |\n",
       "| chr1         | 265404    | 265405    | +            | -1           |\n",
       "| chr1         | 266854    | 266855    | +            | -1           |\n",
       "| ...          | ...       | ...       | ...          | ...          |\n",
       "| chrY         | 1596449   | 1596450   | -            | -1           |\n",
       "| chrY         | 2500649   | 2500650   | -            | -1           |\n",
       "| chrY         | 2500975   | 2500976   | -            | 4            |\n",
       "| chrY         | 57201427  | 57201428  | -            | -1           |\n",
       "+--------------+-----------+-----------+--------------+--------------+\n",
       "Stranded PyRanges object has 111,408 rows and 5 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tool_gtfs.loc[(tool_gtfs['tool'] == display_tool) & (tool_gtfs['tissue'] == display_tissue), 'CAGE_clusters'].values[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RNA-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = \"/project/hfa_work/ENCODE/code/snakemake-pipeline/results/rnaseq/depth_{side}_{tool}_{tissue}.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Imported depth information for flair aorta. Calculating ratio...\n",
      "Imported depth information for flair brain. Calculating ratio...\n",
      "Imported depth information for flair colon. Calculating ratio...\n",
      "Imported depth information for flair heart. Calculating ratio...\n",
      "Imported depth information for flair lung. Calculating ratio...\n",
      "Imported depth information for flair muscle. Calculating ratio...\n",
      "Imported depth information for isotools aorta. Calculating ratio...\n",
      "Imported depth information for isotools brain. Calculating ratio...\n",
      "Imported depth information for isotools colon. Calculating ratio...\n",
      "Imported depth information for isotools heart. Calculating ratio...\n",
      "Imported depth information for isotools lung. Calculating ratio...\n",
      "Imported depth information for isotools muscle. Calculating ratio...\n"
     ]
    }
   ],
   "source": [
    "def tss_ratio(tool, tissue, tss_sites):\n",
    "    # reads in output from samtools depth\n",
    "    # chr, pos, [depth for each sample]\n",
    "    inside_depth_df = pd.read_csv(path.format(side='inside', tool=tool, tissue=tissue),\n",
    "                               sep='\\t',\n",
    "                               header=None,\n",
    "                               names=[\"Chromosome\", \"Start\"] + [f\"inside_sample_{i}\" for i in range(len(metadata[metadata.group == tissue]))])\n",
    "    inside_depth_df['Start'] = inside_depth_df['Start'] - 1\n",
    "    inside_depth_df['End'] = inside_depth_df['Start'] + 1\n",
    "    inside_depth = pr.PyRanges(inside_depth_df)\n",
    "    tss_inside_ranges = tss_sites.assign(\"TSS\", lambda x: x.Start).extend({\"3\": window_padding})\n",
    "    tss_inside_joined = tss_inside_ranges.join(inside_depth, strandedness=False, how='left')\n",
    "\n",
    "    sample_columns_inside = tss_inside_joined.columns[tss_inside_joined.columns.str.startswith('inside_sample_')].values.tolist()\n",
    "    agg = {col: \"mean\" for col in sample_columns_inside}\n",
    "    agg[\"TSS\"] = \"first\"\n",
    "    agg[\"CAGE_score\"] = \"first\"\n",
    "    tss_inside_joined = tss_inside_joined.cluster(slack=-1).boundaries(\"Cluster\", agg=agg).drop(\"Cluster\")\n",
    "    tss_inside_joined = tss_inside_joined.assign(\"Start\", lambda x: x.TSS).assign(\"End\", lambda x: x.TSS + 1).drop(\"TSS\")\n",
    "\n",
    "    outside_depth_df = pd.read_csv(path.format(side='outside', tool=tool, tissue=tissue),\n",
    "                                 sep='\\t',\n",
    "                                 header=None,\n",
    "                                 names=[\"Chromosome\", \"Start\"] + [f\"outside_sample_{i}\" for i in range(len(metadata[metadata.group == tissue]))])\n",
    "    outside_depth_df['Start'] = outside_depth_df['Start'] - 1\n",
    "    outside_depth_df['End'] = outside_depth_df['Start'] + 1\n",
    "    outside_depth = pr.PyRanges(outside_depth_df)\n",
    "    tss_outside_ranges = tss_sites.assign(\"TSS\", lambda x: x.Start).extend({\"5\": window_padding})\n",
    "    tss_outside_joined = tss_outside_ranges.join(outside_depth, strandedness=False, how='left')\n",
    "\n",
    "    sample_columns_outside = tss_outside_joined.columns[tss_outside_joined.columns.str.startswith('outside_sample_')].values.tolist()\n",
    "    agg = {col: \"mean\" for col in sample_columns_outside}\n",
    "    agg[\"TSS\"] = \"first\"\n",
    "    agg[\"CAGE_score\"] = \"first\"\n",
    "    tss_outside_joined = tss_outside_joined.cluster(slack=-1).boundaries(\"Cluster\", agg=agg).drop(\"Cluster\")\n",
    "    tss_outside_joined = tss_outside_joined.assign(\"Start\", lambda x: x.TSS).assign(\"End\", lambda x: x.TSS + 1).drop(\"TSS\")\n",
    "\n",
    "    print(f\"Imported depth information for {tool} {tissue}. Calculating ratio...\")\n",
    "\n",
    "    combined = tss_inside_joined.join(tss_outside_joined, strandedness=False)\n",
    "    combined = combined.drop([\"Start_b\", \"End_b\", \"Strand_b\", \"CAGE_score_b\"])\n",
    "    combined = combined.apply(lambda x: sample_tss_ratios(tissue, x))\n",
    "    combined = combined.apply(lambda x: mean_tss_ratio(tissue, x))\n",
    "    return combined\n",
    "\n",
    "def sample_tss_ratios(tissue, df):\n",
    "    for i in range(len(metadata[metadata.group == tissue])):\n",
    "        df[f\"tss_ratio_sample_{i}\"] = ((df[f\"inside_sample_{i}\"] + 0.01) / (df[f\"outside_sample_{i}\"] + 0.01)).where(df[f\"inside_sample_{i}\"].gt(0), 0)\n",
    "        df.drop([f\"inside_sample_{i}\", f\"outside_sample_{i}\"], axis=1, inplace=True)\n",
    "    return df\n",
    "\n",
    "def mean_tss_ratio(tissue, df):\n",
    "    num_samples = len(metadata[metadata.group == tissue])\n",
    "    ratio_columns = [f\"tss_ratio_sample_{i}\" for i in range(num_samples)]\n",
    "    df[\"TSS_ratio\"] = df[ratio_columns].mean(axis=1)\n",
    "    return df\n",
    "\n",
    "tool_gtfs['CAGE_and_TSS'] = tool_gtfs.apply(lambda x: tss_ratio(x['tool'], x['tissue'], x['CAGE_clusters']), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>CAGE_score</th>\n",
       "      <th>tss_ratio_sample_0</th>\n",
       "      <th>tss_ratio_sample_1</th>\n",
       "      <th>tss_ratio_sample_2</th>\n",
       "      <th>tss_ratio_sample_3</th>\n",
       "      <th>TSS_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>175387</td>\n",
       "      <td>175388</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "      <td>6.039859</td>\n",
       "      <td>6.604772</td>\n",
       "      <td>4.858931</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.375890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>197700</td>\n",
       "      <td>197701</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.076653</td>\n",
       "      <td>0.026797</td>\n",
       "      <td>0.374133</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.119396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>265404</td>\n",
       "      <td>265405</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>266854</td>\n",
       "      <td>266855</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.631859</td>\n",
       "      <td>16.833974</td>\n",
       "      <td>0.247599</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.428358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>271731</td>\n",
       "      <td>271732</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107794</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1452928</td>\n",
       "      <td>1452929</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107795</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1596449</td>\n",
       "      <td>1596450</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107796</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2500649</td>\n",
       "      <td>2500650</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107797</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2500975</td>\n",
       "      <td>2500976</td>\n",
       "      <td>-</td>\n",
       "      <td>4</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107798</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57201427</td>\n",
       "      <td>57201428</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>107799 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +6    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| chr1         | 175387    | 175388    | +            | ...   |\n",
       "| chr1         | 197700    | 197701    | +            | ...   |\n",
       "| chr1         | 265404    | 265405    | +            | ...   |\n",
       "| chr1         | 266854    | 266855    | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| chrY         | 1596449   | 1596450   | -            | ...   |\n",
       "| chrY         | 2500649   | 2500650   | -            | ...   |\n",
       "| chrY         | 2500975   | 2500976   | -            | ...   |\n",
       "| chrY         | 57201427  | 57201428  | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 107,799 rows and 10 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "6 hidden columns: CAGE_score, tss_ratio_sample_0, tss_ratio_sample_1, ... (+ 3 more.)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tool_gtfs.loc[(tool_gtfs['tool'] == display_tool) & (tool_gtfs['tissue'] == display_tissue), 'CAGE_and_TSS'].values[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tool</th>\n",
       "      <th>tissue</th>\n",
       "      <th>num_transcripts</th>\n",
       "      <th>CAGE support absolute</th>\n",
       "      <th>CAGE support relative</th>\n",
       "      <th>TSS ratio absolute</th>\n",
       "      <th>TSS ratio relative</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>flair</td>\n",
       "      <td>aorta</td>\n",
       "      <td>86590</td>\n",
       "      <td>25377</td>\n",
       "      <td>0.293071</td>\n",
       "      <td>40896</td>\n",
       "      <td>0.472295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>flair</td>\n",
       "      <td>brain</td>\n",
       "      <td>107799</td>\n",
       "      <td>31745</td>\n",
       "      <td>0.294483</td>\n",
       "      <td>47542</td>\n",
       "      <td>0.441024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>flair</td>\n",
       "      <td>colon</td>\n",
       "      <td>40834</td>\n",
       "      <td>22250</td>\n",
       "      <td>0.544889</td>\n",
       "      <td>23491</td>\n",
       "      <td>0.575280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>flair</td>\n",
       "      <td>heart</td>\n",
       "      <td>113023</td>\n",
       "      <td>31154</td>\n",
       "      <td>0.275643</td>\n",
       "      <td>82224</td>\n",
       "      <td>0.727498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>flair</td>\n",
       "      <td>lung</td>\n",
       "      <td>93837</td>\n",
       "      <td>31754</td>\n",
       "      <td>0.338395</td>\n",
       "      <td>53391</td>\n",
       "      <td>0.568976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>flair</td>\n",
       "      <td>muscle</td>\n",
       "      <td>25858</td>\n",
       "      <td>17584</td>\n",
       "      <td>0.680022</td>\n",
       "      <td>16323</td>\n",
       "      <td>0.631255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>isotools</td>\n",
       "      <td>aorta</td>\n",
       "      <td>18326</td>\n",
       "      <td>11557</td>\n",
       "      <td>0.630634</td>\n",
       "      <td>10582</td>\n",
       "      <td>0.577431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>isotools</td>\n",
       "      <td>brain</td>\n",
       "      <td>24205</td>\n",
       "      <td>17787</td>\n",
       "      <td>0.734848</td>\n",
       "      <td>12214</td>\n",
       "      <td>0.504606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>isotools</td>\n",
       "      <td>colon</td>\n",
       "      <td>15890</td>\n",
       "      <td>12688</td>\n",
       "      <td>0.798490</td>\n",
       "      <td>10242</td>\n",
       "      <td>0.644556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>isotools</td>\n",
       "      <td>heart</td>\n",
       "      <td>41235</td>\n",
       "      <td>23789</td>\n",
       "      <td>0.576913</td>\n",
       "      <td>24775</td>\n",
       "      <td>0.600825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>isotools</td>\n",
       "      <td>lung</td>\n",
       "      <td>28464</td>\n",
       "      <td>19402</td>\n",
       "      <td>0.681633</td>\n",
       "      <td>16929</td>\n",
       "      <td>0.594751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>isotools</td>\n",
       "      <td>muscle</td>\n",
       "      <td>11886</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.837372</td>\n",
       "      <td>8224</td>\n",
       "      <td>0.691906</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        tool  tissue  num_transcripts  CAGE support absolute  \\\n",
       "0      flair   aorta            86590                  25377   \n",
       "1      flair   brain           107799                  31745   \n",
       "2      flair   colon            40834                  22250   \n",
       "3      flair   heart           113023                  31154   \n",
       "4      flair    lung            93837                  31754   \n",
       "5      flair  muscle            25858                  17584   \n",
       "6   isotools   aorta            18326                  11557   \n",
       "7   isotools   brain            24205                  17787   \n",
       "8   isotools   colon            15890                  12688   \n",
       "9   isotools   heart            41235                  23789   \n",
       "10  isotools    lung            28464                  19402   \n",
       "11  isotools  muscle            11886                   9953   \n",
       "\n",
       "    CAGE support relative  TSS ratio absolute  TSS ratio relative  \n",
       "0                0.293071               40896            0.472295  \n",
       "1                0.294483               47542            0.441024  \n",
       "2                0.544889               23491            0.575280  \n",
       "3                0.275643               82224            0.727498  \n",
       "4                0.338395               53391            0.568976  \n",
       "5                0.680022               16323            0.631255  \n",
       "6                0.630634               10582            0.577431  \n",
       "7                0.734848               12214            0.504606  \n",
       "8                0.798490               10242            0.644556  \n",
       "9                0.576913               24775            0.600825  \n",
       "10               0.681633               16929            0.594751  \n",
       "11               0.837372                8224            0.691906  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tool_gtfs['num_transcripts'] = tool_gtfs['CAGE_and_TSS'].apply(lambda x: x.length)\n",
    "tool_gtfs['CAGE support absolute'] = tool_gtfs['CAGE_and_TSS'].apply(lambda x: x[x.CAGE_score > 0].length)\n",
    "tool_gtfs['CAGE support relative'] = tool_gtfs['CAGE support absolute'] / tool_gtfs['num_transcripts']\n",
    "tool_gtfs['TSS ratio absolute'] = tool_gtfs['CAGE_and_TSS'].apply(lambda x: x[x.TSS_ratio > 1.5].length)\n",
    "tool_gtfs['TSS ratio relative'] = tool_gtfs['TSS ratio absolute'] / tool_gtfs['num_transcripts']\n",
    "\n",
    "output = tool_gtfs.loc[:, ['tool', 'tissue', 'num_transcripts', 'CAGE support absolute', 'CAGE support relative', 'TSS ratio absolute', 'TSS ratio relative']]\n",
    "output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Export"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "!mkdir -p /project/hfa_work/ENCODE/code/snakemake-pipeline/results/cage/\n",
    "with open('/project/hfa_work/ENCODE/code/snakemake-pipeline/results/cage/support.pkl', 'wb') as f:\n",
    "    pickle.dump(tool_gtfs, f)\n",
    "output.to_csv('/project/hfa_work/ENCODE/code/snakemake-pipeline/results/cage/support.tsv', sep='\\t', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "isodev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
