{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Hide FutureWarning\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import pyranges as pr\n",
    "\n",
    "path = '/path/to/data'\n",
    "alignment_path = 'alignment_v45'\n",
    "genome_file = 'GRCh38.p14.genome.fa'\n",
    "genome_path = os.path.join(path, 'gencode_human/version_45', genome_file)\n",
    "window_padding = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_file = 'reads/metadata_tissue.tsv'\n",
    "metadata = pd.read_csv(os.path.join(path, metadata_file), sep='\\t')\n",
    "tissues = metadata['group'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['aorta', 'brain', 'colon', 'heart', 'lung', 'muscle']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CAGE_path = \"/path/to/code/snakemake-pipeline/resources/CAGE/\"\n",
    "# filter out tissues with no corresponding CAGE data\n",
    "tissues = [tissue for tissue in tissues if os.path.exists(os.path.join(CAGE_path, f'{tissue}.bed'))]\n",
    "tissues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# isotools_path = '/path/to/code/notebooks/results/isotools_'\n",
    "fofn_path = '/path/to/code/snakemake-pipeline/results/tissues_gtf.fofn'\n",
    "\n",
    "tool_gtfs = pd.read_csv(fofn_path, sep='\\t', header=0)\n",
    "# check that there are no duplicates of tool + tissue\n",
    "assert tool_gtfs.groupby(['tool', 'tissue']).size().max() == 1\n",
    "# filter for tissues with CAGE data\n",
    "tool_gtfs = tool_gtfs[tool_gtfs['tissue'].isin(tissues)]\n",
    "tool_gtfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "display_tool = 'FLAIR'\n",
    "display_tissue = 'brain'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "CAGE_data = {}\n",
    "for tissue in tissues:\n",
    "    CAGE_data[tissue] = pr.read_bed(os.path.join(CAGE_path, f'{tissue}.bed'))\n",
    "\n",
    "tool_gtfs['pr'] = tool_gtfs['gtf'].apply(lambda x: pr.read_gtf(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>transcript</td>\n",
       "      <td>175387</td>\n",
       "      <td>175644</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:175000</td>\n",
       "      <td>m64109_220203_225821/159646893/ccs</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>175387</td>\n",
       "      <td>175644</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:175000</td>\n",
       "      <td>m64109_220203_225821/159646893/ccs</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>transcript</td>\n",
       "      <td>197700</td>\n",
       "      <td>199762</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:197000</td>\n",
       "      <td>m64109_220205_045735/17237446/ccs</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>197700</td>\n",
       "      <td>199762</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:197000</td>\n",
       "      <td>m64109_220205_045735/17237446/ccs</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>transcript</td>\n",
       "      <td>265404</td>\n",
       "      <td>267777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:265000</td>\n",
       "      <td>m64109_220205_045735/39125266/ccs</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>821086</th>\n",
       "      <td>chrY</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>307731</td>\n",
       "      <td>307880</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292358.1</td>\n",
       "      <td>m64109_220203_225821/137103082/ccs</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>821087</th>\n",
       "      <td>chrY</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>311418</td>\n",
       "      <td>311627</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292358.1</td>\n",
       "      <td>m64109_220203_225821/137103082/ccs</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>821088</th>\n",
       "      <td>chrY</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>312765</td>\n",
       "      <td>315020</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292358.1</td>\n",
       "      <td>m64109_220203_225821/137103082/ccs</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>821089</th>\n",
       "      <td>chrY</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>316913</td>\n",
       "      <td>317051</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292358.1</td>\n",
       "      <td>m64109_220203_225821/137103082/ccs</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>821090</th>\n",
       "      <td>chrY</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>318438</td>\n",
       "      <td>318796</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292358.1</td>\n",
       "      <td>m64109_220203_225821/137103082/ccs</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>821091 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +6    |\n",
       "| (category)   | (object)   | (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | FLAIR      | transcript   | 175387    | 175644    | ...   |\n",
       "| chr1         | FLAIR      | exon         | 175387    | 175644    | ...   |\n",
       "| chr1         | FLAIR      | transcript   | 197700    | 199762    | ...   |\n",
       "| chr1         | FLAIR      | exon         | 197700    | 199762    | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chrY         | FLAIR      | exon         | 311418    | 311627    | ...   |\n",
       "| chrY         | FLAIR      | exon         | 312765    | 315020    | ...   |\n",
       "| chrY         | FLAIR      | exon         | 316913    | 317051    | ...   |\n",
       "| chrY         | FLAIR      | exon         | 318438    | 318796    | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 821,091 rows and 11 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "6 hidden columns: Score, Strand, Frame, gene_id, transcript_id, exon_number"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tool_gtfs.loc[(tool_gtfs['tool'] == display_tool) & (tool_gtfs['tissue'] == display_tissue), 'pr'].values[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Find TSS based on strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19826 duplicate positions dropped\n",
      "32504 duplicate positions dropped\n",
      "21453 duplicate positions dropped\n",
      "95768 duplicate positions dropped\n",
      "50988 duplicate positions dropped\n",
      "17123 duplicate positions dropped\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>175387</td>\n",
       "      <td>175388</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>197700</td>\n",
       "      <td>197701</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>265404</td>\n",
       "      <td>265405</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>266854</td>\n",
       "      <td>266855</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>271731</td>\n",
       "      <td>271732</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111404</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1392112</td>\n",
       "      <td>1392113</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111405</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1452928</td>\n",
       "      <td>1452929</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111406</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2500975</td>\n",
       "      <td>2500976</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111407</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2500649</td>\n",
       "      <td>2500650</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111408</th>\n",
       "      <td>chrY</td>\n",
       "      <td>318795</td>\n",
       "      <td>318796</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>111409 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+\n",
       "| Chromosome   | Start     | End       | Strand       |\n",
       "| (category)   | (int64)   | (int64)   | (category)   |\n",
       "|--------------+-----------+-----------+--------------|\n",
       "| chr1         | 175387    | 175388    | +            |\n",
       "| chr1         | 197700    | 197701    | +            |\n",
       "| chr1         | 265404    | 265405    | +            |\n",
       "| chr1         | 266854    | 266855    | +            |\n",
       "| ...          | ...       | ...       | ...          |\n",
       "| chrY         | 1452928   | 1452929   | -            |\n",
       "| chrY         | 2500975   | 2500976   | -            |\n",
       "| chrY         | 2500649   | 2500650   | -            |\n",
       "| chrY         | 318795    | 318796    | -            |\n",
       "+--------------+-----------+-----------+--------------+\n",
       "Stranded PyRanges object has 111,409 rows and 4 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# tss_cat = [\"5' fragment\", \"novel exonic TSS\", \"novel intronic TSS\"]\n",
    "\n",
    "def tss_from_transcriptome(transcriptome):\n",
    "    transcriptome = transcriptome[transcriptome.Feature == 'transcript']\n",
    "    starts = transcriptome[transcriptome.Strand == '+']\n",
    "    starts.End = starts.Start + 1\n",
    "    ends = transcriptome[transcriptome.Strand == '-']\n",
    "    ends.Start = ends.End - 1\n",
    "    tss_sites = pr.concat([starts, ends]).drop_duplicate_positions()\n",
    "    print(pr.concat([starts, ends]).df.shape[0] - tss_sites.df.shape[0], \"duplicate positions dropped\")\n",
    "    return tss_sites.drop()\n",
    "\n",
    "tool_gtfs['tss_sites'] = tool_gtfs['pr'].apply(tss_from_transcriptome)\n",
    "tool_gtfs.loc[(tool_gtfs['tool'] == display_tool) & (tool_gtfs['tissue'] == display_tissue), 'tss_sites'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629179</td>\n",
       "      <td>629180</td>\n",
       "      <td>chr1:629179..629180,+</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629210</td>\n",
       "      <td>629211</td>\n",
       "      <td>chr1:629210..629211,+</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629212</td>\n",
       "      <td>629213</td>\n",
       "      <td>chr1:629212..629213,+</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629620</td>\n",
       "      <td>629621</td>\n",
       "      <td>chr1:629620..629621,+</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629625</td>\n",
       "      <td>629626</td>\n",
       "      <td>chr1:629625..629626,+</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3059674</th>\n",
       "      <td>chrY</td>\n",
       "      <td>20782681</td>\n",
       "      <td>20782682</td>\n",
       "      <td>chrY:20782681..20782682,-</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3059675</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21303006</td>\n",
       "      <td>21303007</td>\n",
       "      <td>chrY:21303006..21303007,-</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3059676</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25626516</td>\n",
       "      <td>25626517</td>\n",
       "      <td>chrY:25626516..25626517,-</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3059677</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26401653</td>\n",
       "      <td>26401654</td>\n",
       "      <td>chrY:26401653..26401654,-</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3059678</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26493515</td>\n",
       "      <td>26493516</td>\n",
       "      <td>chrY:26493515..26493516,-</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3059679 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+---------------------------+-------+\n",
       "| Chromosome   | Start     | End       | Name                      | +2    |\n",
       "| (category)   | (int64)   | (int64)   | (object)                  | ...   |\n",
       "|--------------+-----------+-----------+---------------------------+-------|\n",
       "| chr1         | 629179    | 629180    | chr1:629179..629180,+     | ...   |\n",
       "| chr1         | 629210    | 629211    | chr1:629210..629211,+     | ...   |\n",
       "| chr1         | 629212    | 629213    | chr1:629212..629213,+     | ...   |\n",
       "| chr1         | 629620    | 629621    | chr1:629620..629621,+     | ...   |\n",
       "| ...          | ...       | ...       | ...                       | ...   |\n",
       "| chrY         | 21303006  | 21303007  | chrY:21303006..21303007,- | ...   |\n",
       "| chrY         | 25626516  | 25626517  | chrY:25626516..25626517,- | ...   |\n",
       "| chrY         | 26401653  | 26401654  | chrY:26401653..26401654,- | ...   |\n",
       "| chrY         | 26493515  | 26493516  | chrY:26493515..26493516,- | ...   |\n",
       "+--------------+-----------+-----------+---------------------------+-------+\n",
       "Stranded PyRanges object has 3,059,679 rows and 6 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "2 hidden columns: Score, Strand"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CAGE_data[display_tissue]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629079</td>\n",
       "      <td>629280</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629110</td>\n",
       "      <td>629311</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629112</td>\n",
       "      <td>629313</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629520</td>\n",
       "      <td>629721</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629525</td>\n",
       "      <td>629726</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3059674</th>\n",
       "      <td>chrY</td>\n",
       "      <td>20782581</td>\n",
       "      <td>20782782</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3059675</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21302906</td>\n",
       "      <td>21303107</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3059676</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25626416</td>\n",
       "      <td>25626617</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3059677</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26401553</td>\n",
       "      <td>26401754</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3059678</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26493415</td>\n",
       "      <td>26493616</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3059679 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-----------+--------------+\n",
       "| Chromosome   | Start     | End       | Score     | Strand       |\n",
       "| (category)   | (int64)   | (int64)   | (int64)   | (category)   |\n",
       "|--------------+-----------+-----------+-----------+--------------|\n",
       "| chr1         | 629079    | 629280    | 1         | +            |\n",
       "| chr1         | 629110    | 629311    | 1         | +            |\n",
       "| chr1         | 629112    | 629313    | 1         | +            |\n",
       "| chr1         | 629520    | 629721    | 1         | +            |\n",
       "| ...          | ...       | ...       | ...       | ...          |\n",
       "| chrY         | 21302906  | 21303107  | 1         | -            |\n",
       "| chrY         | 25626416  | 25626617  | 1         | -            |\n",
       "| chrY         | 26401553  | 26401754  | 1         | -            |\n",
       "| chrY         | 26493415  | 26493616  | 1         | -            |\n",
       "+--------------+-----------+-----------+-----------+--------------+\n",
       "Stranded PyRanges object has 3,059,679 rows and 5 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extend positions by 20bp in each direction\n",
    "CAGE_windows = {tissue: CAGE_data[tissue][['Score']].extend(window_padding) for tissue in tissues}\n",
    "CAGE_windows[display_tissue]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------+-----------+-----------+--------------+\n",
      "| Chromosome   | Start     | End       | Strand       |\n",
      "| (category)   | (int64)   | (int64)   | (category)   |\n",
      "|--------------+-----------+-----------+--------------|\n",
      "| chr1         | 19963     | 19964     | +            |\n",
      "| chr1         | 195837    | 195838    | +            |\n",
      "| chr1         | 296931    | 296932    | +            |\n",
      "| chr1         | 493489    | 493490    | +            |\n",
      "| ...          | ...       | ...       | ...          |\n",
      "| chrY         | 1392112   | 1392113   | -            |\n",
      "| chrY         | 1452928   | 1452929   | -            |\n",
      "| chrY         | 2500975   | 2500976   | -            |\n",
      "| chrY         | 2500649   | 2500650   | -            |\n",
      "+--------------+-----------+-----------+--------------+\n",
      "Stranded PyRanges object has 89,221 rows and 4 columns from 25 chromosomes.\n",
      "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
      "+--------------+-----------+-----------+--------------+\n",
      "| Chromosome   | Start     | End       | Strand       |\n",
      "| (category)   | (int64)   | (int64)   | (category)   |\n",
      "|--------------+-----------+-----------+--------------|\n",
      "| chr1         | 175387    | 175388    | +            |\n",
      "| chr1         | 197700    | 197701    | +            |\n",
      "| chr1         | 265404    | 265405    | +            |\n",
      "| chr1         | 266854    | 266855    | +            |\n",
      "| ...          | ...       | ...       | ...          |\n",
      "| chrY         | 1452928   | 1452929   | -            |\n",
      "| chrY         | 2500975   | 2500976   | -            |\n",
      "| chrY         | 2500649   | 2500650   | -            |\n",
      "| chrY         | 318795    | 318796    | -            |\n",
      "+--------------+-----------+-----------+--------------+\n",
      "Stranded PyRanges object has 111,409 rows and 4 columns from 25 chromosomes.\n",
      "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
      "+--------------+-----------+-----------+--------------+\n",
      "| Chromosome   | Start     | End       | Strand       |\n",
      "| (category)   | (int64)   | (int64)   | (category)   |\n",
      "|--------------+-----------+-----------+--------------|\n",
      "| chr1         | 189895    | 189896    | +            |\n",
      "| chr1         | 629306    | 629307    | +            |\n",
      "| chr1         | 629503    | 629504    | +            |\n",
      "| chr1         | 629639    | 629640    | +            |\n",
      "| ...          | ...       | ...       | ...          |\n",
      "| chrY         | 19744938  | 19744939  | -            |\n",
      "| chrY         | 19744710  | 19744711  | -            |\n",
      "| chrY         | 20575361  | 20575362  | -            |\n",
      "| chrY         | 2500649   | 2500650   | -            |\n",
      "+--------------+-----------+-----------+--------------+\n",
      "Stranded PyRanges object has 43,343 rows and 4 columns from 25 chromosomes.\n",
      "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
      "+--------------+-----------+-----------+--------------+\n",
      "| Chromosome   | Start     | End       | Strand       |\n",
      "| (category)   | (int64)   | (int64)   | (category)   |\n",
      "|--------------+-----------+-----------+--------------|\n",
      "| chr1         | 197661    | 197662    | +            |\n",
      "| chr1         | 521285    | 521286    | +            |\n",
      "| chr1         | 629306    | 629307    | +            |\n",
      "| chr1         | 629502    | 629503    | +            |\n",
      "| ...          | ...       | ...       | ...          |\n",
      "| chrY         | 20575496  | 20575497  | -            |\n",
      "| chrY         | 25394718  | 25394719  | -            |\n",
      "| chrY         | 19710605  | 19710606  | -            |\n",
      "| chrY         | 19733725  | 19733726  | -            |\n",
      "+--------------+-----------+-----------+--------------+\n",
      "Stranded PyRanges object has 120,390 rows and 4 columns from 25 chromosomes.\n",
      "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
      "+--------------+-----------+-----------+--------------+\n",
      "| Chromosome   | Start     | End       | Strand       |\n",
      "| (category)   | (int64)   | (int64)   | (category)   |\n",
      "|--------------+-----------+-----------+--------------|\n",
      "| chr1         | 28588     | 28589     | +            |\n",
      "| chr1         | 123792    | 123793    | +            |\n",
      "| chr1         | 136648    | 136649    | +            |\n",
      "| chr1         | 174959    | 174960    | +            |\n",
      "| ...          | ...       | ...       | ...          |\n",
      "| chrY         | 386954    | 386955    | -            |\n",
      "| chrY         | 2502804   | 2502805   | -            |\n",
      "| chrY         | 12421586  | 12421587  | -            |\n",
      "| chrY         | 20575361  | 20575362  | -            |\n",
      "+--------------+-----------+-----------+--------------+\n",
      "Stranded PyRanges object has 99,081 rows and 4 columns from 25 chromosomes.\n",
      "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
      "+--------------+-----------+-----------+--------------+\n",
      "| Chromosome   | Start     | End       | Strand       |\n",
      "| (category)   | (int64)   | (int64)   | (category)   |\n",
      "|--------------+-----------+-----------+--------------|\n",
      "| chr1         | 629306    | 629307    | +            |\n",
      "| chr1         | 630780    | 630781    | +            |\n",
      "| chr1         | 633952    | 633953    | +            |\n",
      "| chr1         | 853824    | 853825    | +            |\n",
      "| ...          | ...       | ...       | ...          |\n",
      "| chrY         | 374043    | 374044    | -            |\n",
      "| chrY         | 1392112   | 1392113   | -            |\n",
      "| chrY         | 2500649   | 2500650   | -            |\n",
      "| chrY         | 2500959   | 2500960   | -            |\n",
      "+--------------+-----------+-----------+--------------+\n",
      "Stranded PyRanges object has 27,721 rows and 4 columns from 25 chromosomes.\n",
      "For printing, the PyRanges was sorted on Chromosome and Strand.\n"
     ]
    }
   ],
   "source": [
    "def add_CAGE_Support(tss_sites, CAGE_windows):\n",
    "    merged = tss_sites.join(CAGE_windows, strandedness='same', how='left')\n",
    "    clustered = merged.cluster(slack=-1).boundaries(\"Cluster\", agg={\"Score\": \"sum\"})\n",
    "    return clustered\n",
    "\n",
    "tool_gtfs['CAGE_clusters'] = tool_gtfs.apply(lambda x: add_CAGE_Support(x['tss_sites'], CAGE_windows[x['tissue']]), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Score</th>\n",
       "      <th>Cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>175387</td>\n",
       "      <td>175388</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>197700</td>\n",
       "      <td>197701</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>265404</td>\n",
       "      <td>265405</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>266854</td>\n",
       "      <td>266855</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>271731</td>\n",
       "      <td>271732</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111404</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1452928</td>\n",
       "      <td>1452929</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>111405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111405</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1596449</td>\n",
       "      <td>1596450</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>111406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111406</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2500649</td>\n",
       "      <td>2500650</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>111407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111407</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2500975</td>\n",
       "      <td>2500976</td>\n",
       "      <td>-</td>\n",
       "      <td>4</td>\n",
       "      <td>111408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111408</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57201427</td>\n",
       "      <td>57201428</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>111409</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>111409 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-----------+-----------+\n",
       "| Chromosome   | Start     | End       | Strand       | Score     | Cluster   |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | (int64)   | (int64)   |\n",
       "|--------------+-----------+-----------+--------------+-----------+-----------|\n",
       "| chr1         | 175387    | 175388    | +            | -1        | 1         |\n",
       "| chr1         | 197700    | 197701    | +            | -1        | 2         |\n",
       "| chr1         | 265404    | 265405    | +            | -1        | 3         |\n",
       "| chr1         | 266854    | 266855    | +            | -1        | 4         |\n",
       "| ...          | ...       | ...       | ...          | ...       | ...       |\n",
       "| chrY         | 1596449   | 1596450   | -            | -1        | 111406    |\n",
       "| chrY         | 2500649   | 2500650   | -            | -1        | 111407    |\n",
       "| chrY         | 2500975   | 2500976   | -            | 4         | 111408    |\n",
       "| chrY         | 57201427  | 57201428  | -            | -1        | 111409    |\n",
       "+--------------+-----------+-----------+--------------+-----------+-----------+\n",
       "Stranded PyRanges object has 111,409 rows and 6 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tool_gtfs.loc[(tool_gtfs['tool'] == display_tool) & (tool_gtfs['tissue'] == display_tissue), 'CAGE_clusters'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tool</th>\n",
       "      <th>tissue</th>\n",
       "      <th>num_transcripts</th>\n",
       "      <th>CAGE support absolute</th>\n",
       "      <th>CAGE support relative</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FLAIR</td>\n",
       "      <td>aorta</td>\n",
       "      <td>89221</td>\n",
       "      <td>27938</td>\n",
       "      <td>0.313133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FLAIR</td>\n",
       "      <td>brain</td>\n",
       "      <td>111409</td>\n",
       "      <td>35306</td>\n",
       "      <td>0.316904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FLAIR</td>\n",
       "      <td>colon</td>\n",
       "      <td>43343</td>\n",
       "      <td>24719</td>\n",
       "      <td>0.570311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FLAIR</td>\n",
       "      <td>heart</td>\n",
       "      <td>120390</td>\n",
       "      <td>37958</td>\n",
       "      <td>0.315292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FLAIR</td>\n",
       "      <td>lung</td>\n",
       "      <td>99081</td>\n",
       "      <td>36854</td>\n",
       "      <td>0.371958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>FLAIR</td>\n",
       "      <td>muscle</td>\n",
       "      <td>27721</td>\n",
       "      <td>19427</td>\n",
       "      <td>0.700804</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    tool  tissue  num_transcripts  CAGE support absolute  \\\n",
       "0  FLAIR   aorta            89221                  27938   \n",
       "1  FLAIR   brain           111409                  35306   \n",
       "2  FLAIR   colon            43343                  24719   \n",
       "3  FLAIR   heart           120390                  37958   \n",
       "4  FLAIR    lung            99081                  36854   \n",
       "5  FLAIR  muscle            27721                  19427   \n",
       "\n",
       "   CAGE support relative  \n",
       "0               0.313133  \n",
       "1               0.316904  \n",
       "2               0.570311  \n",
       "3               0.315292  \n",
       "4               0.371958  \n",
       "5               0.700804  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tool_gtfs['num_transcripts'] = tool_gtfs['CAGE_clusters'].apply(lambda x: x.length)\n",
    "tool_gtfs['CAGE support absolute'] = tool_gtfs['CAGE_clusters'].apply(lambda x: x[x.Score > 0].length)\n",
    "tool_gtfs['CAGE support relative'] = tool_gtfs['CAGE support absolute'] / tool_gtfs['num_transcripts']\n",
    "\n",
    "tool_gtfs.loc[:, ['tool', 'tissue', 'num_transcripts', 'CAGE support absolute', 'CAGE support relative']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# for tissue in tissues:\n",
    "#     print('\\n\\n', tissue, '\\n')\n",
    "#     # count numbers of each novelty combination\n",
    "#     print(CAGE_clusters[tissue].df.groupby(tss_cat).size())\n",
    "#     print('---')\n",
    "#     print(CAGE_clusters[tissue][CAGE_clusters[tissue].Score > 0].df.groupby(tss_cat).size())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for tissue in tissues:\n",
    "#     print('\\n\\n', tissue, '\\n')\n",
    "#     # Percentages in each group\n",
    "#     print(CAGE_clusters[tissue][CAGE_clusters[tissue].Score > 0].df.groupby(tss_cat).size() / CAGE_clusters[tissue].df.groupby(tss_cat).size())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "isodev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
