{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('/home/lankenau/isotools/src')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "STRICT = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import math\n",
    "\n",
    "path = '/project/hfa_work/ENCODE/data'\n",
    "alignment_path = 'alignment_v45'\n",
    "genome_file = 'GRCh38.p14.genome.fa'\n",
    "genome_path = os.path.join(path, 'gencode_human/version_45', genome_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample ID</th>\n",
       "      <th>organism</th>\n",
       "      <th>individual</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>developmental stage</th>\n",
       "      <th>disease</th>\n",
       "      <th>group</th>\n",
       "      <th>Biosample term name</th>\n",
       "      <th>Description</th>\n",
       "      <th>Biosample summary</th>\n",
       "      <th>file</th>\n",
       "      <th>file size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCSR700EBI</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W61</td>\n",
       "      <td>female</td>\n",
       "      <td>41 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>aorta</td>\n",
       "      <td>aorta</td>\n",
       "      <td>W61 aorta ENC4_88</td>\n",
       "      <td>Homo sapiens aorta tissue female adult (41 years)</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>616290149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCSR425HFS</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W73</td>\n",
       "      <td>female</td>\n",
       "      <td>59 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>aorta</td>\n",
       "      <td>aorta</td>\n",
       "      <td>ENC4_103 W73 aorta</td>\n",
       "      <td>Homo sapiens aorta tissue female adult (59 years)</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>1168326040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCSR463IDK</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>E5194210</td>\n",
       "      <td>female</td>\n",
       "      <td>79 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>brain</td>\n",
       "      <td>dorsolateral prefrontal cortex</td>\n",
       "      <td>RUSH brain E5194210</td>\n",
       "      <td>Homo sapiens dorsolateral prefrontal cortex ti...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>1145642918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCSR205QMF</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>E5115600</td>\n",
       "      <td>female</td>\n",
       "      <td>88 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>brain</td>\n",
       "      <td>dorsolateral prefrontal cortex</td>\n",
       "      <td>RUSH brain E5115600</td>\n",
       "      <td>Homo sapiens dorsolateral prefrontal cortex ti...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>452782844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCSR169YNI</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>E7461192</td>\n",
       "      <td>female</td>\n",
       "      <td>90 or above years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>brain</td>\n",
       "      <td>dorsolateral prefrontal cortex</td>\n",
       "      <td>RUSH brain E7461192</td>\n",
       "      <td>Homo sapiens dorsolateral prefrontal cortex ti...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>866488131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENCSR094NFM</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>E4368365</td>\n",
       "      <td>female</td>\n",
       "      <td>90 or above years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>brain</td>\n",
       "      <td>dorsolateral prefrontal cortex</td>\n",
       "      <td>RUSH brain E4368365</td>\n",
       "      <td>Homo sapiens dorsolateral prefrontal cortex ti...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>193641719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENCSR997RFW</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W76</td>\n",
       "      <td>male</td>\n",
       "      <td>40 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>colon</td>\n",
       "      <td>mucosa of descending colon</td>\n",
       "      <td>ENC4_77 W76 mucosa of descending colon</td>\n",
       "      <td>Homo sapiens mucosa of descending colon tissue...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>800980029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENCSR450GAR</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W73</td>\n",
       "      <td>female</td>\n",
       "      <td>59 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>colon</td>\n",
       "      <td>left colon</td>\n",
       "      <td>W73 left colon ENC4_87</td>\n",
       "      <td>Homo sapiens left colon tissue female adult (5...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>730236367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENCSR984OAE</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W76</td>\n",
       "      <td>male</td>\n",
       "      <td>40 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>heart</td>\n",
       "      <td>heart right ventricle</td>\n",
       "      <td>ENC4_52 W76 heart right ventricle</td>\n",
       "      <td>Homo sapiens heart right ventricle tissue male...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>686400256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENCSR994YZY</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W76</td>\n",
       "      <td>male</td>\n",
       "      <td>40 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>heart</td>\n",
       "      <td>heart left ventricle</td>\n",
       "      <td>ENC4_53 W76 heart left ventricle</td>\n",
       "      <td>Homo sapiens heart left ventricle tissue male ...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>658616334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENCSR728TXV</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W76</td>\n",
       "      <td>male</td>\n",
       "      <td>40 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>heart</td>\n",
       "      <td>right cardiac atrium</td>\n",
       "      <td>W76 right cardiac atrium ENC4_84</td>\n",
       "      <td>Homo sapiens right cardiac atrium tissue male ...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>1070292006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENCSR549ELD</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W61</td>\n",
       "      <td>female</td>\n",
       "      <td>41 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>heart</td>\n",
       "      <td>cardiac septum</td>\n",
       "      <td>W61 cardiac septum ENC4_92</td>\n",
       "      <td>Homo sapiens cardiac septum tissue female adul...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>966929649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENCSR782LGT</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W72</td>\n",
       "      <td>female</td>\n",
       "      <td>46 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>heart</td>\n",
       "      <td>heart right ventricle</td>\n",
       "      <td>ENC4_50 W72 heart right ventricle</td>\n",
       "      <td>Homo sapiens heart right ventricle tissue fema...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>823409465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENCSR514YQN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W72</td>\n",
       "      <td>female</td>\n",
       "      <td>46 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>heart</td>\n",
       "      <td>right cardiac atrium</td>\n",
       "      <td>ENC4_59 W72 right cardiac atrium</td>\n",
       "      <td>Homo sapiens right cardiac atrium tissue femal...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>531001643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENCSR700XDQ</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W72</td>\n",
       "      <td>female</td>\n",
       "      <td>46 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>heart</td>\n",
       "      <td>heart left ventricle</td>\n",
       "      <td>ENC4_49 W72 heart left ventricle</td>\n",
       "      <td>Homo sapiens heart left ventricle tissue femal...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>754519472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENCSR424QFN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W73</td>\n",
       "      <td>female</td>\n",
       "      <td>59 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>heart</td>\n",
       "      <td>left cardiac atrium</td>\n",
       "      <td>ENC4_60 W73 left cardiac atrium</td>\n",
       "      <td>Homo sapiens left cardiac atrium tissue female...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>513336683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENCSR575LWI</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W73</td>\n",
       "      <td>female</td>\n",
       "      <td>59 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>heart</td>\n",
       "      <td>heart left ventricle</td>\n",
       "      <td>ENC4_33 W73 heart left ventricle</td>\n",
       "      <td>Homo sapiens heart left ventricle tissue femal...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>1596671766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENCSR329ZQG</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W73</td>\n",
       "      <td>female</td>\n",
       "      <td>59 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>heart</td>\n",
       "      <td>heart right ventricle</td>\n",
       "      <td>ENC4_51 W73 heart right ventricle</td>\n",
       "      <td>Homo sapiens heart right ventricle tissue fema...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>1141023769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENCSR553SVP</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W73</td>\n",
       "      <td>female</td>\n",
       "      <td>59 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>heart</td>\n",
       "      <td>right cardiac atrium</td>\n",
       "      <td>ENC4_36 W73 cardiac right atrium</td>\n",
       "      <td>Homo sapiens right cardiac atrium tissue femal...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>1728935144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENCSR777CCI</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W78</td>\n",
       "      <td>male</td>\n",
       "      <td>60 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>heart</td>\n",
       "      <td>left ventricle myocardium superior</td>\n",
       "      <td>ENC4_54 W78 left ventricle myocardium superior</td>\n",
       "      <td>Homo sapiens left ventricle myocardium superio...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>834422860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENCSR435UUS</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W78</td>\n",
       "      <td>male</td>\n",
       "      <td>60 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>heart</td>\n",
       "      <td>right cardiac atrium</td>\n",
       "      <td>ENC4_58 W78 right cardiac atrium</td>\n",
       "      <td>Homo sapiens right cardiac atrium tissue male ...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>485270428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ENCSR899GAP</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W78</td>\n",
       "      <td>male</td>\n",
       "      <td>60 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>heart</td>\n",
       "      <td>Right ventricle myocardium superior</td>\n",
       "      <td>ENC4_56 W78 right ventricle myocardium superior</td>\n",
       "      <td>Homo sapiens Right ventricle myocardium superi...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>1106765207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENCSR591OZR</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W78</td>\n",
       "      <td>male</td>\n",
       "      <td>60 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>heart</td>\n",
       "      <td>Right ventricle myocardium inferior</td>\n",
       "      <td>ENC4_57 W78 right ventricle myocardium inferior</td>\n",
       "      <td>Homo sapiens Right ventricle myocardium inferi...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>429678973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENCSR786FLO</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W78</td>\n",
       "      <td>male</td>\n",
       "      <td>60 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>heart</td>\n",
       "      <td>left ventricle myocardium inferior</td>\n",
       "      <td>ENC4_55 W78 left ventricle myocardium inferior</td>\n",
       "      <td>Homo sapiens left ventricle myocardium inferio...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>975804751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ENCSR261GOA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W73</td>\n",
       "      <td>female</td>\n",
       "      <td>59 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>lung</td>\n",
       "      <td>lower lobe of left lung</td>\n",
       "      <td>ENC4_80 W73 lower lobe of left lung</td>\n",
       "      <td>Homo sapiens lower lobe of left lung tissue fe...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>1508113793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ENCSR096QUP</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W78</td>\n",
       "      <td>male</td>\n",
       "      <td>60 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>lung</td>\n",
       "      <td>upper lobe of right lung</td>\n",
       "      <td>ENC4_82 W78 upper lobe of right lung</td>\n",
       "      <td>Homo sapiens upper lobe of right lung tissue m...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>1173492279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ENCSR986WKB</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W78</td>\n",
       "      <td>male</td>\n",
       "      <td>60 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>lung</td>\n",
       "      <td>lower lobe of left lung</td>\n",
       "      <td>ENC4_39 W78 left lobe of lung</td>\n",
       "      <td>Homo sapiens lower lobe of left lung tissue ma...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>496759987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ENCSR323XND</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W78</td>\n",
       "      <td>male</td>\n",
       "      <td>60 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>lung</td>\n",
       "      <td>lower lobe of right lung</td>\n",
       "      <td>W78 lower lobe of right lung ENC4_83</td>\n",
       "      <td>Homo sapiens lower lobe of right lung tissue m...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>1051682756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ENCSR803QWH</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W61</td>\n",
       "      <td>female</td>\n",
       "      <td>41 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>muscle</td>\n",
       "      <td>psoas muscle</td>\n",
       "      <td>ENC4_46 W61 psoas #2</td>\n",
       "      <td>Homo sapiens psoas muscle tissue female adult ...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>472845478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ENCSR243DYK</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W80</td>\n",
       "      <td>female</td>\n",
       "      <td>61 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>muscle</td>\n",
       "      <td>psoas muscle</td>\n",
       "      <td>W80 psoas muscle ENC4_86</td>\n",
       "      <td>Homo sapiens psoas muscle tissue female adult ...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>1029550105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ENCSR477HHG</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W73</td>\n",
       "      <td>female</td>\n",
       "      <td>59 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>ovary</td>\n",
       "      <td>ovary</td>\n",
       "      <td>ENC4_71 W73 ovary</td>\n",
       "      <td>Homo sapiens ovary tissue female adult (59 years)</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>1189739402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>ENCSR507WZC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W80</td>\n",
       "      <td>female</td>\n",
       "      <td>61 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>ovary</td>\n",
       "      <td>ovary</td>\n",
       "      <td>ENC4_72 W80 ovary</td>\n",
       "      <td>Homo sapiens ovary tissue female adult (61 years)</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>951175430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ENCSR853YZN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W71</td>\n",
       "      <td>female</td>\n",
       "      <td>47 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>vessel</td>\n",
       "      <td>posterior vena cava</td>\n",
       "      <td>W71 posterior vena cava ENC4_101</td>\n",
       "      <td>Homo sapiens posterior vena cava tissue female...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>344068378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>ENCSR138TAS</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>W73</td>\n",
       "      <td>female</td>\n",
       "      <td>59 years</td>\n",
       "      <td>adult</td>\n",
       "      <td>normal</td>\n",
       "      <td>vessel</td>\n",
       "      <td>posterior vena cava</td>\n",
       "      <td>ENC4_102 W73 posterior vena cava</td>\n",
       "      <td>Homo sapiens posterior vena cava tissue female...</td>\n",
       "      <td>/project/hfa_work/yalan/golong/encode_tissue_d...</td>\n",
       "      <td>699897074</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample ID      organism individual     sex                age  \\\n",
       "0   ENCSR700EBI  Homo sapiens        W61  female           41 years   \n",
       "1   ENCSR425HFS  Homo sapiens        W73  female           59 years   \n",
       "2   ENCSR463IDK  Homo sapiens   E5194210  female           79 years   \n",
       "3   ENCSR205QMF  Homo sapiens   E5115600  female           88 years   \n",
       "4   ENCSR169YNI  Homo sapiens   E7461192  female  90 or above years   \n",
       "5   ENCSR094NFM  Homo sapiens   E4368365  female  90 or above years   \n",
       "6   ENCSR997RFW  Homo sapiens        W76    male           40 years   \n",
       "7   ENCSR450GAR  Homo sapiens        W73  female           59 years   \n",
       "8   ENCSR984OAE  Homo sapiens        W76    male           40 years   \n",
       "9   ENCSR994YZY  Homo sapiens        W76    male           40 years   \n",
       "10  ENCSR728TXV  Homo sapiens        W76    male           40 years   \n",
       "11  ENCSR549ELD  Homo sapiens        W61  female           41 years   \n",
       "12  ENCSR782LGT  Homo sapiens        W72  female           46 years   \n",
       "13  ENCSR514YQN  Homo sapiens        W72  female           46 years   \n",
       "14  ENCSR700XDQ  Homo sapiens        W72  female           46 years   \n",
       "15  ENCSR424QFN  Homo sapiens        W73  female           59 years   \n",
       "16  ENCSR575LWI  Homo sapiens        W73  female           59 years   \n",
       "17  ENCSR329ZQG  Homo sapiens        W73  female           59 years   \n",
       "18  ENCSR553SVP  Homo sapiens        W73  female           59 years   \n",
       "19  ENCSR777CCI  Homo sapiens        W78    male           60 years   \n",
       "20  ENCSR435UUS  Homo sapiens        W78    male           60 years   \n",
       "21  ENCSR899GAP  Homo sapiens        W78    male           60 years   \n",
       "22  ENCSR591OZR  Homo sapiens        W78    male           60 years   \n",
       "23  ENCSR786FLO  Homo sapiens        W78    male           60 years   \n",
       "24  ENCSR261GOA  Homo sapiens        W73  female           59 years   \n",
       "25  ENCSR096QUP  Homo sapiens        W78    male           60 years   \n",
       "26  ENCSR986WKB  Homo sapiens        W78    male           60 years   \n",
       "27  ENCSR323XND  Homo sapiens        W78    male           60 years   \n",
       "28  ENCSR803QWH  Homo sapiens        W61  female           41 years   \n",
       "29  ENCSR243DYK  Homo sapiens        W80  female           61 years   \n",
       "30  ENCSR477HHG  Homo sapiens        W73  female           59 years   \n",
       "31  ENCSR507WZC  Homo sapiens        W80  female           61 years   \n",
       "32  ENCSR853YZN  Homo sapiens        W71  female           47 years   \n",
       "33  ENCSR138TAS  Homo sapiens        W73  female           59 years   \n",
       "\n",
       "   developmental stage disease   group                  Biosample term name  \\\n",
       "0                adult  normal   aorta                                aorta   \n",
       "1                adult  normal   aorta                                aorta   \n",
       "2                adult  normal   brain       dorsolateral prefrontal cortex   \n",
       "3                adult  normal   brain       dorsolateral prefrontal cortex   \n",
       "4                adult  normal   brain       dorsolateral prefrontal cortex   \n",
       "5                adult  normal   brain       dorsolateral prefrontal cortex   \n",
       "6                adult  normal   colon           mucosa of descending colon   \n",
       "7                adult  normal   colon                           left colon   \n",
       "8                adult  normal   heart                heart right ventricle   \n",
       "9                adult  normal   heart                 heart left ventricle   \n",
       "10               adult  normal   heart                 right cardiac atrium   \n",
       "11               adult  normal   heart                       cardiac septum   \n",
       "12               adult  normal   heart                heart right ventricle   \n",
       "13               adult  normal   heart                 right cardiac atrium   \n",
       "14               adult  normal   heart                 heart left ventricle   \n",
       "15               adult  normal   heart                  left cardiac atrium   \n",
       "16               adult  normal   heart                 heart left ventricle   \n",
       "17               adult  normal   heart                heart right ventricle   \n",
       "18               adult  normal   heart                 right cardiac atrium   \n",
       "19               adult  normal   heart   left ventricle myocardium superior   \n",
       "20               adult  normal   heart                 right cardiac atrium   \n",
       "21               adult  normal   heart  Right ventricle myocardium superior   \n",
       "22               adult  normal   heart  Right ventricle myocardium inferior   \n",
       "23               adult  normal   heart   left ventricle myocardium inferior   \n",
       "24               adult  normal    lung              lower lobe of left lung   \n",
       "25               adult  normal    lung             upper lobe of right lung   \n",
       "26               adult  normal    lung              lower lobe of left lung   \n",
       "27               adult  normal    lung             lower lobe of right lung   \n",
       "28               adult  normal  muscle                         psoas muscle   \n",
       "29               adult  normal  muscle                         psoas muscle   \n",
       "30               adult  normal   ovary                                ovary   \n",
       "31               adult  normal   ovary                                ovary   \n",
       "32               adult  normal  vessel                  posterior vena cava   \n",
       "33               adult  normal  vessel                  posterior vena cava   \n",
       "\n",
       "                                        Description  \\\n",
       "0                                 W61 aorta ENC4_88   \n",
       "1                                ENC4_103 W73 aorta   \n",
       "2                               RUSH brain E5194210   \n",
       "3                               RUSH brain E5115600   \n",
       "4                               RUSH brain E7461192   \n",
       "5                               RUSH brain E4368365   \n",
       "6            ENC4_77 W76 mucosa of descending colon   \n",
       "7                            W73 left colon ENC4_87   \n",
       "8                 ENC4_52 W76 heart right ventricle   \n",
       "9                  ENC4_53 W76 heart left ventricle   \n",
       "10                 W76 right cardiac atrium ENC4_84   \n",
       "11                       W61 cardiac septum ENC4_92   \n",
       "12                ENC4_50 W72 heart right ventricle   \n",
       "13                 ENC4_59 W72 right cardiac atrium   \n",
       "14                 ENC4_49 W72 heart left ventricle   \n",
       "15                  ENC4_60 W73 left cardiac atrium   \n",
       "16                 ENC4_33 W73 heart left ventricle   \n",
       "17                ENC4_51 W73 heart right ventricle   \n",
       "18                 ENC4_36 W73 cardiac right atrium   \n",
       "19   ENC4_54 W78 left ventricle myocardium superior   \n",
       "20                 ENC4_58 W78 right cardiac atrium   \n",
       "21  ENC4_56 W78 right ventricle myocardium superior   \n",
       "22  ENC4_57 W78 right ventricle myocardium inferior   \n",
       "23   ENC4_55 W78 left ventricle myocardium inferior   \n",
       "24              ENC4_80 W73 lower lobe of left lung   \n",
       "25             ENC4_82 W78 upper lobe of right lung   \n",
       "26                    ENC4_39 W78 left lobe of lung   \n",
       "27             W78 lower lobe of right lung ENC4_83   \n",
       "28                             ENC4_46 W61 psoas #2   \n",
       "29                         W80 psoas muscle ENC4_86   \n",
       "30                                ENC4_71 W73 ovary   \n",
       "31                                ENC4_72 W80 ovary   \n",
       "32                 W71 posterior vena cava ENC4_101   \n",
       "33                 ENC4_102 W73 posterior vena cava   \n",
       "\n",
       "                                    Biosample summary  \\\n",
       "0   Homo sapiens aorta tissue female adult (41 years)   \n",
       "1   Homo sapiens aorta tissue female adult (59 years)   \n",
       "2   Homo sapiens dorsolateral prefrontal cortex ti...   \n",
       "3   Homo sapiens dorsolateral prefrontal cortex ti...   \n",
       "4   Homo sapiens dorsolateral prefrontal cortex ti...   \n",
       "5   Homo sapiens dorsolateral prefrontal cortex ti...   \n",
       "6   Homo sapiens mucosa of descending colon tissue...   \n",
       "7   Homo sapiens left colon tissue female adult (5...   \n",
       "8   Homo sapiens heart right ventricle tissue male...   \n",
       "9   Homo sapiens heart left ventricle tissue male ...   \n",
       "10  Homo sapiens right cardiac atrium tissue male ...   \n",
       "11  Homo sapiens cardiac septum tissue female adul...   \n",
       "12  Homo sapiens heart right ventricle tissue fema...   \n",
       "13  Homo sapiens right cardiac atrium tissue femal...   \n",
       "14  Homo sapiens heart left ventricle tissue femal...   \n",
       "15  Homo sapiens left cardiac atrium tissue female...   \n",
       "16  Homo sapiens heart left ventricle tissue femal...   \n",
       "17  Homo sapiens heart right ventricle tissue fema...   \n",
       "18  Homo sapiens right cardiac atrium tissue femal...   \n",
       "19  Homo sapiens left ventricle myocardium superio...   \n",
       "20  Homo sapiens right cardiac atrium tissue male ...   \n",
       "21  Homo sapiens Right ventricle myocardium superi...   \n",
       "22  Homo sapiens Right ventricle myocardium inferi...   \n",
       "23  Homo sapiens left ventricle myocardium inferio...   \n",
       "24  Homo sapiens lower lobe of left lung tissue fe...   \n",
       "25  Homo sapiens upper lobe of right lung tissue m...   \n",
       "26  Homo sapiens lower lobe of left lung tissue ma...   \n",
       "27  Homo sapiens lower lobe of right lung tissue m...   \n",
       "28  Homo sapiens psoas muscle tissue female adult ...   \n",
       "29  Homo sapiens psoas muscle tissue female adult ...   \n",
       "30  Homo sapiens ovary tissue female adult (59 years)   \n",
       "31  Homo sapiens ovary tissue female adult (61 years)   \n",
       "32  Homo sapiens posterior vena cava tissue female...   \n",
       "33  Homo sapiens posterior vena cava tissue female...   \n",
       "\n",
       "                                                 file   file size  \n",
       "0   /project/hfa_work/yalan/golong/encode_tissue_d...   616290149  \n",
       "1   /project/hfa_work/yalan/golong/encode_tissue_d...  1168326040  \n",
       "2   /project/hfa_work/yalan/golong/encode_tissue_d...  1145642918  \n",
       "3   /project/hfa_work/yalan/golong/encode_tissue_d...   452782844  \n",
       "4   /project/hfa_work/yalan/golong/encode_tissue_d...   866488131  \n",
       "5   /project/hfa_work/yalan/golong/encode_tissue_d...   193641719  \n",
       "6   /project/hfa_work/yalan/golong/encode_tissue_d...   800980029  \n",
       "7   /project/hfa_work/yalan/golong/encode_tissue_d...   730236367  \n",
       "8   /project/hfa_work/yalan/golong/encode_tissue_d...   686400256  \n",
       "9   /project/hfa_work/yalan/golong/encode_tissue_d...   658616334  \n",
       "10  /project/hfa_work/yalan/golong/encode_tissue_d...  1070292006  \n",
       "11  /project/hfa_work/yalan/golong/encode_tissue_d...   966929649  \n",
       "12  /project/hfa_work/yalan/golong/encode_tissue_d...   823409465  \n",
       "13  /project/hfa_work/yalan/golong/encode_tissue_d...   531001643  \n",
       "14  /project/hfa_work/yalan/golong/encode_tissue_d...   754519472  \n",
       "15  /project/hfa_work/yalan/golong/encode_tissue_d...   513336683  \n",
       "16  /project/hfa_work/yalan/golong/encode_tissue_d...  1596671766  \n",
       "17  /project/hfa_work/yalan/golong/encode_tissue_d...  1141023769  \n",
       "18  /project/hfa_work/yalan/golong/encode_tissue_d...  1728935144  \n",
       "19  /project/hfa_work/yalan/golong/encode_tissue_d...   834422860  \n",
       "20  /project/hfa_work/yalan/golong/encode_tissue_d...   485270428  \n",
       "21  /project/hfa_work/yalan/golong/encode_tissue_d...  1106765207  \n",
       "22  /project/hfa_work/yalan/golong/encode_tissue_d...   429678973  \n",
       "23  /project/hfa_work/yalan/golong/encode_tissue_d...   975804751  \n",
       "24  /project/hfa_work/yalan/golong/encode_tissue_d...  1508113793  \n",
       "25  /project/hfa_work/yalan/golong/encode_tissue_d...  1173492279  \n",
       "26  /project/hfa_work/yalan/golong/encode_tissue_d...   496759987  \n",
       "27  /project/hfa_work/yalan/golong/encode_tissue_d...  1051682756  \n",
       "28  /project/hfa_work/yalan/golong/encode_tissue_d...   472845478  \n",
       "29  /project/hfa_work/yalan/golong/encode_tissue_d...  1029550105  \n",
       "30  /project/hfa_work/yalan/golong/encode_tissue_d...  1189739402  \n",
       "31  /project/hfa_work/yalan/golong/encode_tissue_d...   951175430  \n",
       "32  /project/hfa_work/yalan/golong/encode_tissue_d...   344068378  \n",
       "33  /project/hfa_work/yalan/golong/encode_tissue_d...   699897074  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_file = 'reads/metadata_tissue.tsv'\n",
    "metadata = pd.read_csv(os.path.join(path, metadata_file), sep='\\t')\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:This is isotools version 0.3.5rc10\n"
     ]
    }
   ],
   "source": [
    "import logging\n",
    "from isotools import Transcriptome\n",
    "from isotools import __version__ as isotools_version\n",
    "# set up logging\n",
    "logging.basicConfig(format='%(levelname)s:%(message)s', level=logging.INFO)\n",
    "logger=logging.getLogger('isotools')\n",
    "logger.info(f'This is isotools version {isotools_version}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:importing reference from gff3 file /project/hfa_work/ENCODE/data/gencode_human/version_45/gencode.v45.chr_patch_hapl_scaff.annotation_sorted.gff3.gz\n",
      "  0%|          | 0.00/81.7M [00:00<?, ?B/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████▉| 81.7M/81.7M [00:46<00:00, 1.84MB/s]\n",
      "INFO:skipped the following categories: {'CDS', 'three_prime_UTR', 'five_prime_UTR', 'stop_codon_redefined_as_selenocysteine'}\n"
     ]
    }
   ],
   "source": [
    "annotation_file=os.path.join(path, 'gencode_human/version_45', 'gencode.v45.chr_patch_hapl_scaff.annotation_sorted.gff3.gz')\n",
    "#create the IsoTools transcriptome object from the reference annotation\n",
    "isoseq=Transcriptome.from_reference(annotation_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:adding sample ENCSR700EBI from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF144KHH_aligned.bam\n",
      "  0%|          | 0.00/2.19M [00:00<?, ?reads/s, chr=chr1]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 2.19M/2.19M [02:55<00:00, 12.5kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 2720 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 950420 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 123726 chimeric alignments with less than 2 reads\n",
      "INFO:imported 1106210 nonchimeric reads (including  1377 chained chimeric alignments) and 9024 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR425HFS from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF902BIU_aligned.bam\n",
      "100%|██████████| 2.74M/2.74M [05:05<00:00, 8.96kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 9838 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 667992 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 50134 chimeric alignments with less than 2 reads\n",
      "INFO:imported 2013267 nonchimeric reads (including  5697 chained chimeric alignments) and 2918 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR463IDK from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF838DFB_aligned.bam\n",
      "100%|██████████| 2.79M/2.79M [04:50<00:00, 9.59kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 3181 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 591511 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 139059 chimeric alignments with less than 2 reads\n",
      "INFO:imported 2035784 nonchimeric reads (including  5726 chained chimeric alignments) and 18875 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR205QMF from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF206TQZ_aligned.bam\n",
      "100%|██████████| 1.37M/1.37M [02:06<00:00, 10.9kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 1746 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 330427 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 69758 chimeric alignments with less than 2 reads\n",
      "INFO:imported 963304 nonchimeric reads (including  2063 chained chimeric alignments) and 8488 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR169YNI from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF827DUW_aligned.bam\n",
      "[W::hts_idx_load3] The index file is older than the data file: /project/hfa_work/ENCODE/data/alignment_v45/ENCFF827DUW_aligned.bam.bai\n",
      "100%|██████████| 2.18M/2.18M [03:31<00:00, 10.3kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 3079 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 483534 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 85750 chimeric alignments with less than 2 reads\n",
      "INFO:imported 1593059 nonchimeric reads (including  3699 chained chimeric alignments) and 11852 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR094NFM from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF260AWP_aligned.bam\n",
      "100%|██████████| 748k/748k [01:11<00:00, 10.4kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 791 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 206335 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 33463 chimeric alignments with less than 2 reads\n",
      "INFO:imported 501838 nonchimeric reads (including  843 chained chimeric alignments) and 5193 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR997RFW from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF387HPO_aligned.bam\n",
      "100%|██████████| 2.33M/2.33M [04:29<00:00, 8.62kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 3145 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 887153 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 145755 chimeric alignments with less than 2 reads\n",
      "INFO:imported 1272092 nonchimeric reads (including  1519 chained chimeric alignments) and 16637 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR450GAR from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF245MBY_aligned.bam\n",
      "100%|██████████| 1.85M/1.85M [03:17<00:00, 9.37kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 2720 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 513797 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 51473 chimeric alignments with less than 2 reads\n",
      "INFO:imported 1282606 nonchimeric reads (including  562 chained chimeric alignments) and 3025 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR984OAE from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF425VDL_aligned.bam\n",
      "100%|██████████| 1.45M/1.45M [04:00<00:00, 6.03kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 2203 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 291367 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 95187 chimeric alignments with less than 2 reads\n",
      "INFO:imported 1053851 nonchimeric reads (including  2259 chained chimeric alignments) and 9911 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR994YZY from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF602MAI_aligned.bam\n",
      "[W::hts_idx_load3] The index file is older than the data file: /project/hfa_work/ENCODE/data/alignment_v45/ENCFF602MAI_aligned.bam.bai\n",
      "100%|██████████| 1.37M/1.37M [03:24<00:00, 6.70kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 2692 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 268238 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 93435 chimeric alignments with less than 2 reads\n",
      "INFO:imported 992724 nonchimeric reads (including  2131 chained chimeric alignments) and 11205 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR728TXV from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF905RVF_aligned.bam\n",
      "100%|██████████| 2.63M/2.63M [05:19<00:00, 8.22kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 2804 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 724705 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 143915 chimeric alignments with less than 2 reads\n",
      "INFO:imported 1734681 nonchimeric reads (including  3399 chained chimeric alignments) and 19105 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR549ELD from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF680XXE_aligned.bam\n",
      "100%|██████████| 2.72M/2.72M [05:33<00:00, 8.15kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 3792 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 785179 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 152297 chimeric alignments with less than 2 reads\n",
      "INFO:imported 1757108 nonchimeric reads (including  2713 chained chimeric alignments) and 21115 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR782LGT from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF793PGJ_aligned.bam\n",
      "[W::hts_idx_load3] The index file is older than the data file: /project/hfa_work/ENCODE/data/alignment_v45/ENCFF793PGJ_aligned.bam.bai\n",
      "100%|██████████| 1.74M/1.74M [04:44<00:00, 6.12kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 2083 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 328541 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 111472 chimeric alignments with less than 2 reads\n",
      "INFO:imported 1282692 nonchimeric reads (including  3003 chained chimeric alignments) and 14812 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR514YQN from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF722JJS_aligned.bam\n",
      "100%|██████████| 1.02M/1.02M [02:23<00:00, 7.14kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 2023 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 188431 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 69189 chimeric alignments with less than 2 reads\n",
      "INFO:imported 753768 nonchimeric reads (including  1581 chained chimeric alignments) and 8863 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR700XDQ from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF429JUP_aligned.bam\n",
      "100%|██████████| 1.66M/1.66M [04:28<00:00, 6.18kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 3879 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 305396 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 117260 chimeric alignments with less than 2 reads\n",
      "INFO:imported 1216655 nonchimeric reads (including  2566 chained chimeric alignments) and 17612 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR424QFN from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF920VXE_aligned.bam\n",
      "100%|██████████| 948k/948k [03:12<00:00, 4.93kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 1451 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 171896 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 11522 chimeric alignments with less than 2 reads\n",
      "INFO:imported 764050 nonchimeric reads (including  1445 chained chimeric alignments) and 314 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR575LWI from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF537NCV_aligned.bam\n",
      "100%|██████████| 3.18M/3.18M [11:24<00:00, 4.65kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 13098 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 503833 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 29215 chimeric alignments with less than 2 reads\n",
      "INFO:imported 2636899 nonchimeric reads (including  4045 chained chimeric alignments) and 1769 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR329ZQG from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF615FIC_aligned.bam\n",
      "100%|██████████| 2.02M/2.02M [07:57<00:00, 4.22kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 11637 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 314308 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 28229 chimeric alignments with less than 2 reads\n",
      "INFO:imported 1661473 nonchimeric reads (including  2765 chained chimeric alignments) and 1811 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR553SVP from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF899MTI_aligned.bam\n",
      "100%|██████████| 3.02M/3.02M [12:06<00:00, 4.16kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 13606 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 510083 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 38198 chimeric alignments with less than 2 reads\n",
      "INFO:imported 2461011 nonchimeric reads (including  5207 chained chimeric alignments) and 2814 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR777CCI from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF624IQY_aligned.bam\n",
      "100%|██████████| 1.95M/1.95M [04:41<00:00, 6.91kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 4429 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 403387 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 124042 chimeric alignments with less than 2 reads\n",
      "INFO:imported 1397321 nonchimeric reads (including  2627 chained chimeric alignments) and 17412 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR435UUS from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF242WRZ_aligned.bam\n",
      "100%|██████████| 1.02M/1.02M [02:22<00:00, 7.14kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 1314 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 212095 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 61978 chimeric alignments with less than 2 reads\n",
      "INFO:imported 732507 nonchimeric reads (including  1377 chained chimeric alignments) and 7194 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR899GAP from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF665LBS_aligned.bam\n",
      "100%|██████████| 2.08M/2.08M [05:15<00:00, 6.60kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 2843 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 355185 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 128291 chimeric alignments with less than 2 reads\n",
      "INFO:imported 1574327 nonchimeric reads (including  3659 chained chimeric alignments) and 21865 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR591OZR from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF738UZJ_aligned.bam\n",
      "100%|██████████| 896k/896k [01:56<00:00, 7.68kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 893 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 179310 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 56594 chimeric alignments with less than 2 reads\n",
      "INFO:imported 652097 nonchimeric reads (including  1280 chained chimeric alignments) and 6756 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR786FLO from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF793CMQ_aligned.bam\n",
      "100%|██████████| 1.82M/1.82M [05:36<00:00, 5.42kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 4288 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 321095 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 119434 chimeric alignments with less than 2 reads\n",
      "INFO:imported 1360397 nonchimeric reads (including  3416 chained chimeric alignments) and 19530 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR261GOA from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF552NVU_aligned.bam\n",
      "100%|██████████| 3.68M/3.68M [10:11<00:00, 6.02kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 14735 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 1061897 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 36504 chimeric alignments with less than 2 reads\n",
      "INFO:imported 2572795 nonchimeric reads (including  4168 chained chimeric alignments) and 1862 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR096QUP from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF934MBW_aligned.bam\n",
      "[W::hts_idx_load3] The index file is older than the data file: /project/hfa_work/ENCODE/data/alignment_v45/ENCFF934MBW_aligned.bam.bai\n",
      "100%|██████████| 2.81M/2.81M [09:26<00:00, 4.95kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 3569 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 997475 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 157066 chimeric alignments with less than 2 reads\n",
      "INFO:imported 1626976 nonchimeric reads (including  2152 chained chimeric alignments) and 18510 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR986WKB from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF341BSQ_aligned.bam\n",
      "100%|██████████| 1.16M/1.16M [03:15<00:00, 5.92kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 2174 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 343544 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 67496 chimeric alignments with less than 2 reads\n",
      "INFO:imported 740372 nonchimeric reads (including  872 chained chimeric alignments) and 4510 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR323XND from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF250IWT_aligned.bam\n",
      "100%|██████████| 3.09M/3.09M [08:08<00:00, 6.32kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 2999 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 1097306 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 162366 chimeric alignments with less than 2 reads\n",
      "INFO:imported 1809228 nonchimeric reads (including  1726 chained chimeric alignments) and 15746 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR803QWH from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF630XEC_aligned.bam\n",
      "100%|██████████| 1.02M/1.02M [03:22<00:00, 5.03kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 2044 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 240631 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 58756 chimeric alignments with less than 2 reads\n",
      "INFO:imported 709483 nonchimeric reads (including  413 chained chimeric alignments) and 7736 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR243DYK from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF750LYC_aligned.bam\n",
      "[W::hts_idx_load3] The index file is older than the data file: /project/hfa_work/ENCODE/data/alignment_v45/ENCFF750LYC_aligned.bam.bai\n",
      "100%|██████████| 2.78M/2.78M [08:00<00:00, 5.78kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 3789 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 839568 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 157013 chimeric alignments with less than 2 reads\n",
      "INFO:imported 1741917 nonchimeric reads (including  958 chained chimeric alignments) and 30931 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR477HHG from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF756AHG_aligned.bam\n",
      "100%|██████████| 3.36M/3.36M [08:33<00:00, 6.54kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 10278 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 1187631 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 30399 chimeric alignments with less than 2 reads\n",
      "INFO:imported 2132071 nonchimeric reads (including  3038 chained chimeric alignments) and 1669 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR507WZC from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF422XLS_aligned.bam\n",
      "100%|██████████| 2.08M/2.08M [06:17<00:00, 5.51kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 3413 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 650188 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 120801 chimeric alignments with less than 2 reads\n",
      "INFO:imported 1290065 nonchimeric reads (including  1850 chained chimeric alignments) and 14977 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR853YZN from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF960KBO_aligned.bam\n",
      "100%|██████████| 1.16M/1.16M [04:15<00:00, 4.54kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 1952 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 498054 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 67339 chimeric alignments with less than 2 reads\n",
      "INFO:imported 582828 nonchimeric reads (including  962 chained chimeric alignments) and 6745 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample ENCSR138TAS from file /project/hfa_work/ENCODE/data/alignment_v45/ENCFF658OZB_aligned.bam\n",
      "100%|██████████| 1.90M/1.90M [04:50<00:00, 6.54kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 5998 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 537609 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "INFO:ignoring 19212 chimeric alignments with less than 2 reads\n",
      "INFO:imported 1337027 nonchimeric reads (including  2934 chained chimeric alignments) and 1228 chimeric reads with coverage of at least 2.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>file</th>\n",
       "      <th>group</th>\n",
       "      <th>nonchimeric_reads</th>\n",
       "      <th>chimeric_reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCSR700EBI</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>aorta</td>\n",
       "      <td>1106210</td>\n",
       "      <td>9024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCSR425HFS</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>aorta</td>\n",
       "      <td>2013267</td>\n",
       "      <td>2918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCSR463IDK</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>brain</td>\n",
       "      <td>2035784</td>\n",
       "      <td>18875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCSR205QMF</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>brain</td>\n",
       "      <td>963304</td>\n",
       "      <td>8488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCSR169YNI</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>brain</td>\n",
       "      <td>1593059</td>\n",
       "      <td>11852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENCSR094NFM</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>brain</td>\n",
       "      <td>501838</td>\n",
       "      <td>5193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENCSR997RFW</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>colon</td>\n",
       "      <td>1272092</td>\n",
       "      <td>16637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENCSR450GAR</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>colon</td>\n",
       "      <td>1282606</td>\n",
       "      <td>3025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENCSR984OAE</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>heart</td>\n",
       "      <td>1053851</td>\n",
       "      <td>9911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENCSR994YZY</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>heart</td>\n",
       "      <td>992724</td>\n",
       "      <td>11205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENCSR728TXV</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>heart</td>\n",
       "      <td>1734681</td>\n",
       "      <td>19105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENCSR549ELD</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>heart</td>\n",
       "      <td>1757108</td>\n",
       "      <td>21115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENCSR782LGT</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>heart</td>\n",
       "      <td>1282692</td>\n",
       "      <td>14812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENCSR514YQN</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>heart</td>\n",
       "      <td>753768</td>\n",
       "      <td>8863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENCSR700XDQ</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>heart</td>\n",
       "      <td>1216655</td>\n",
       "      <td>17612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENCSR424QFN</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>heart</td>\n",
       "      <td>764050</td>\n",
       "      <td>314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENCSR575LWI</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>heart</td>\n",
       "      <td>2636899</td>\n",
       "      <td>1769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENCSR329ZQG</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>heart</td>\n",
       "      <td>1661473</td>\n",
       "      <td>1811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENCSR553SVP</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>heart</td>\n",
       "      <td>2461011</td>\n",
       "      <td>2814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENCSR777CCI</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>heart</td>\n",
       "      <td>1397321</td>\n",
       "      <td>17412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENCSR435UUS</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>heart</td>\n",
       "      <td>732507</td>\n",
       "      <td>7194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ENCSR899GAP</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>heart</td>\n",
       "      <td>1574327</td>\n",
       "      <td>21865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENCSR591OZR</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>heart</td>\n",
       "      <td>652097</td>\n",
       "      <td>6756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENCSR786FLO</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>heart</td>\n",
       "      <td>1360397</td>\n",
       "      <td>19530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ENCSR261GOA</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>lung</td>\n",
       "      <td>2572795</td>\n",
       "      <td>1862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ENCSR096QUP</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>lung</td>\n",
       "      <td>1626976</td>\n",
       "      <td>18510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ENCSR986WKB</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>lung</td>\n",
       "      <td>740372</td>\n",
       "      <td>4510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ENCSR323XND</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>lung</td>\n",
       "      <td>1809228</td>\n",
       "      <td>15746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ENCSR803QWH</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>muscle</td>\n",
       "      <td>709483</td>\n",
       "      <td>7736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ENCSR243DYK</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>muscle</td>\n",
       "      <td>1741917</td>\n",
       "      <td>30931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ENCSR477HHG</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>ovary</td>\n",
       "      <td>2132071</td>\n",
       "      <td>1669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>ENCSR507WZC</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>ovary</td>\n",
       "      <td>1290065</td>\n",
       "      <td>14977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ENCSR853YZN</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>vessel</td>\n",
       "      <td>582828</td>\n",
       "      <td>6745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>ENCSR138TAS</td>\n",
       "      <td>/project/hfa_work/ENCODE/data/alignment_v45/EN...</td>\n",
       "      <td>vessel</td>\n",
       "      <td>1337027</td>\n",
       "      <td>1228</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           name                                               file   group  \\\n",
       "0   ENCSR700EBI  /project/hfa_work/ENCODE/data/alignment_v45/EN...   aorta   \n",
       "1   ENCSR425HFS  /project/hfa_work/ENCODE/data/alignment_v45/EN...   aorta   \n",
       "2   ENCSR463IDK  /project/hfa_work/ENCODE/data/alignment_v45/EN...   brain   \n",
       "3   ENCSR205QMF  /project/hfa_work/ENCODE/data/alignment_v45/EN...   brain   \n",
       "4   ENCSR169YNI  /project/hfa_work/ENCODE/data/alignment_v45/EN...   brain   \n",
       "5   ENCSR094NFM  /project/hfa_work/ENCODE/data/alignment_v45/EN...   brain   \n",
       "6   ENCSR997RFW  /project/hfa_work/ENCODE/data/alignment_v45/EN...   colon   \n",
       "7   ENCSR450GAR  /project/hfa_work/ENCODE/data/alignment_v45/EN...   colon   \n",
       "8   ENCSR984OAE  /project/hfa_work/ENCODE/data/alignment_v45/EN...   heart   \n",
       "9   ENCSR994YZY  /project/hfa_work/ENCODE/data/alignment_v45/EN...   heart   \n",
       "10  ENCSR728TXV  /project/hfa_work/ENCODE/data/alignment_v45/EN...   heart   \n",
       "11  ENCSR549ELD  /project/hfa_work/ENCODE/data/alignment_v45/EN...   heart   \n",
       "12  ENCSR782LGT  /project/hfa_work/ENCODE/data/alignment_v45/EN...   heart   \n",
       "13  ENCSR514YQN  /project/hfa_work/ENCODE/data/alignment_v45/EN...   heart   \n",
       "14  ENCSR700XDQ  /project/hfa_work/ENCODE/data/alignment_v45/EN...   heart   \n",
       "15  ENCSR424QFN  /project/hfa_work/ENCODE/data/alignment_v45/EN...   heart   \n",
       "16  ENCSR575LWI  /project/hfa_work/ENCODE/data/alignment_v45/EN...   heart   \n",
       "17  ENCSR329ZQG  /project/hfa_work/ENCODE/data/alignment_v45/EN...   heart   \n",
       "18  ENCSR553SVP  /project/hfa_work/ENCODE/data/alignment_v45/EN...   heart   \n",
       "19  ENCSR777CCI  /project/hfa_work/ENCODE/data/alignment_v45/EN...   heart   \n",
       "20  ENCSR435UUS  /project/hfa_work/ENCODE/data/alignment_v45/EN...   heart   \n",
       "21  ENCSR899GAP  /project/hfa_work/ENCODE/data/alignment_v45/EN...   heart   \n",
       "22  ENCSR591OZR  /project/hfa_work/ENCODE/data/alignment_v45/EN...   heart   \n",
       "23  ENCSR786FLO  /project/hfa_work/ENCODE/data/alignment_v45/EN...   heart   \n",
       "24  ENCSR261GOA  /project/hfa_work/ENCODE/data/alignment_v45/EN...    lung   \n",
       "25  ENCSR096QUP  /project/hfa_work/ENCODE/data/alignment_v45/EN...    lung   \n",
       "26  ENCSR986WKB  /project/hfa_work/ENCODE/data/alignment_v45/EN...    lung   \n",
       "27  ENCSR323XND  /project/hfa_work/ENCODE/data/alignment_v45/EN...    lung   \n",
       "28  ENCSR803QWH  /project/hfa_work/ENCODE/data/alignment_v45/EN...  muscle   \n",
       "29  ENCSR243DYK  /project/hfa_work/ENCODE/data/alignment_v45/EN...  muscle   \n",
       "30  ENCSR477HHG  /project/hfa_work/ENCODE/data/alignment_v45/EN...   ovary   \n",
       "31  ENCSR507WZC  /project/hfa_work/ENCODE/data/alignment_v45/EN...   ovary   \n",
       "32  ENCSR853YZN  /project/hfa_work/ENCODE/data/alignment_v45/EN...  vessel   \n",
       "33  ENCSR138TAS  /project/hfa_work/ENCODE/data/alignment_v45/EN...  vessel   \n",
       "\n",
       "   nonchimeric_reads chimeric_reads  \n",
       "0            1106210           9024  \n",
       "1            2013267           2918  \n",
       "2            2035784          18875  \n",
       "3             963304           8488  \n",
       "4            1593059          11852  \n",
       "5             501838           5193  \n",
       "6            1272092          16637  \n",
       "7            1282606           3025  \n",
       "8            1053851           9911  \n",
       "9             992724          11205  \n",
       "10           1734681          19105  \n",
       "11           1757108          21115  \n",
       "12           1282692          14812  \n",
       "13            753768           8863  \n",
       "14           1216655          17612  \n",
       "15            764050            314  \n",
       "16           2636899           1769  \n",
       "17           1661473           1811  \n",
       "18           2461011           2814  \n",
       "19           1397321          17412  \n",
       "20            732507           7194  \n",
       "21           1574327          21865  \n",
       "22            652097           6756  \n",
       "23           1360397          19530  \n",
       "24           2572795           1862  \n",
       "25           1626976          18510  \n",
       "26            740372           4510  \n",
       "27           1809228          15746  \n",
       "28            709483           7736  \n",
       "29           1741917          30931  \n",
       "30           2132071           1669  \n",
       "31           1290065          14977  \n",
       "32            582828           6745  \n",
       "33           1337027           1228  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for i, row in metadata.iterrows():\n",
    "    sample_name = row['sample ID']\n",
    "    # file is the full (wrong) path, we just need the filename without the extension\n",
    "    sample_file = os.path.join(path, alignment_path, row['file'].split('/')[-1].split('.')[0] + '_aligned.bam')\n",
    "    if not os.path.exists(sample_file):\n",
    "        logger.error(f'File {sample_file} does not exist')\n",
    "        continue\n",
    "    group = row['group']\n",
    "    isoseq.add_sample_from_bam(fn=sample_file, sample_name=sample_name, group=group, strictness=20 if STRICT else math.inf)\n",
    "isoseq.sample_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 541983/541983 [2:09:49<00:00, 69.58genes/s]  \n",
      "100%|██████████| 541983/541983 [29:24<00:00, 307.23genes/s]\n"
     ]
    }
   ],
   "source": [
    "# compute qc metrics\n",
    "isoseq.add_qc_metrics(genome_path)\n",
    "# add ORF predictions\n",
    "isoseq.add_orf_prediction(genome_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "suffix = '_strict' if STRICT else ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:saving transcriptome to results/isoseq_v45.pkl\n"
     ]
    }
   ],
   "source": [
    "isoseq.save(f'results/isoseq_v45{suffix}.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:writing gtf file to results/isoseq_v45.gtf\n"
     ]
    }
   ],
   "source": [
    "isoseq.write_gtf(f'results/isoseq_v45{suffix}.gtf',\n",
    "                 min_coverage=5, gzip=False, query=\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 541983/541983 [00:20<00:00, 26557.64genes/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>transcript_start</th>\n",
       "      <th>transcript_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_nr</th>\n",
       "      <th>transcript_length</th>\n",
       "      <th>num_exons</th>\n",
       "      <th>exon_starts</th>\n",
       "      <th>...</th>\n",
       "      <th>ovary_sum_coverage</th>\n",
       "      <th>vessel_sum_coverage</th>\n",
       "      <th>aorta_sum_tpm</th>\n",
       "      <th>brain_sum_tpm</th>\n",
       "      <th>colon_sum_tpm</th>\n",
       "      <th>heart_sum_tpm</th>\n",
       "      <th>lung_sum_tpm</th>\n",
       "      <th>muscle_sum_tpm</th>\n",
       "      <th>ovary_sum_tpm</th>\n",
       "      <th>vessel_sum_tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GL000194.1</td>\n",
       "      <td>61662</td>\n",
       "      <td>62921</td>\n",
       "      <td>-</td>\n",
       "      <td>PB_novel_188538</td>\n",
       "      <td>PB_novel_188538</td>\n",
       "      <td>0</td>\n",
       "      <td>1259</td>\n",
       "      <td>1</td>\n",
       "      <td>61662</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>0.320567</td>\n",
       "      <td>0.58893</td>\n",
       "      <td>0.391436</td>\n",
       "      <td>0.771620</td>\n",
       "      <td>0.444486</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.584430</td>\n",
       "      <td>2.083491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GL000194.1</td>\n",
       "      <td>53591</td>\n",
       "      <td>55442</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000277400.1</td>\n",
       "      <td>ENSG00000277400</td>\n",
       "      <td>0</td>\n",
       "      <td>1851</td>\n",
       "      <td>1</td>\n",
       "      <td>53591</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2.243966</td>\n",
       "      <td>0.19631</td>\n",
       "      <td>1.957179</td>\n",
       "      <td>1.043957</td>\n",
       "      <td>1.926105</td>\n",
       "      <td>0.40793</td>\n",
       "      <td>0.876645</td>\n",
       "      <td>1.562618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GL000194.1</td>\n",
       "      <td>53591</td>\n",
       "      <td>115065</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000277400.1</td>\n",
       "      <td>ENSG00000277400</td>\n",
       "      <td>1</td>\n",
       "      <td>2224</td>\n",
       "      <td>3</td>\n",
       "      <td>53591,112791,114985</td>\n",
       "      <td>...</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>3.526232</td>\n",
       "      <td>0.58893</td>\n",
       "      <td>1.957179</td>\n",
       "      <td>1.361683</td>\n",
       "      <td>2.370591</td>\n",
       "      <td>0.40793</td>\n",
       "      <td>3.798797</td>\n",
       "      <td>1.562618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GL000194.1</td>\n",
       "      <td>53591</td>\n",
       "      <td>55503</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000277400.1</td>\n",
       "      <td>ENSG00000277400</td>\n",
       "      <td>2</td>\n",
       "      <td>1127</td>\n",
       "      <td>2</td>\n",
       "      <td>53591,54677</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.320567</td>\n",
       "      <td>0.19631</td>\n",
       "      <td>0.391436</td>\n",
       "      <td>0.226947</td>\n",
       "      <td>0.148162</td>\n",
       "      <td>0.40793</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GL000194.1</td>\n",
       "      <td>53591</td>\n",
       "      <td>115065</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000277400.1</td>\n",
       "      <td>ENSG00000277400</td>\n",
       "      <td>3</td>\n",
       "      <td>1439</td>\n",
       "      <td>4</td>\n",
       "      <td>53591,54677,112791,114985</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.641133</td>\n",
       "      <td>0.39262</td>\n",
       "      <td>0.391436</td>\n",
       "      <td>0.090779</td>\n",
       "      <td>0.148162</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.876645</td>\n",
       "      <td>0.520873</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr  transcript_start  transcript_end strand            gene_id  \\\n",
       "0  GL000194.1             61662           62921      -    PB_novel_188538   \n",
       "1  GL000194.1             53591           55442      -  ENSG00000277400.1   \n",
       "2  GL000194.1             53591          115065      -  ENSG00000277400.1   \n",
       "3  GL000194.1             53591           55503      -  ENSG00000277400.1   \n",
       "4  GL000194.1             53591          115065      -  ENSG00000277400.1   \n",
       "\n",
       "         gene_name  transcript_nr  transcript_length  num_exons  \\\n",
       "0  PB_novel_188538              0               1259          1   \n",
       "1  ENSG00000277400              0               1851          1   \n",
       "2  ENSG00000277400              1               2224          3   \n",
       "3  ENSG00000277400              2               1127          2   \n",
       "4  ENSG00000277400              3               1439          4   \n",
       "\n",
       "                 exon_starts  ... ovary_sum_coverage vessel_sum_coverage  \\\n",
       "0                      61662  ...                  2                   4   \n",
       "1                      53591  ...                  3                   3   \n",
       "2        53591,112791,114985  ...                 13                   3   \n",
       "3                53591,54677  ...                  0                   0   \n",
       "4  53591,54677,112791,114985  ...                  3                   1   \n",
       "\n",
       "  aorta_sum_tpm  brain_sum_tpm  colon_sum_tpm  heart_sum_tpm  lung_sum_tpm  \\\n",
       "0      0.320567        0.58893       0.391436       0.771620      0.444486   \n",
       "1      2.243966        0.19631       1.957179       1.043957      1.926105   \n",
       "2      3.526232        0.58893       1.957179       1.361683      2.370591   \n",
       "3      0.320567        0.19631       0.391436       0.226947      0.148162   \n",
       "4      0.641133        0.39262       0.391436       0.090779      0.148162   \n",
       "\n",
       "   muscle_sum_tpm  ovary_sum_tpm  vessel_sum_tpm  \n",
       "0         0.00000       0.584430        2.083491  \n",
       "1         0.40793       0.876645        1.562618  \n",
       "2         0.40793       3.798797        1.562618  \n",
       "3         0.40793       0.000000        0.000000  \n",
       "4         0.00000       0.876645        0.520873  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# export transcript table with the same filter criteria:\n",
    "transcript_tab=isoseq.transcript_table( groups=isoseq.groups(),tpm=True,coverage=True,\n",
    "                                       min_coverage=5, progress_bar=True,\n",
    "                                       query=\"\")\n",
    "# write to csv file\n",
    "transcript_tab.to_csv(f'results/demonstration_dataset_substantial_transcripts{suffix}.csv',\n",
    "                      index=False, sep='\\t')\n",
    "\n",
    "transcript_tab.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "iso",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
