{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('/home/lankenau/isotools/src')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:This is isotools version 0.3.5rc11\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import logging\n",
    "from isotools import Transcriptome\n",
    "from isotools import __version__ as isotools_version\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "# set up logging\n",
    "logging.basicConfig(format=\"%(levelname)s:%(message)s\", level=logging.INFO)\n",
    "logger = logging.getLogger(\"isotools\")\n",
    "logger.info(f\"This is isotools version {isotools_version}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "isotools_pkl = 'path/to/notebooks/results/flair_isotools.pkl'\n",
    "sqanti_classification = 'path/to/snakemake-pipeline/results/sqanti/flair/qc/aorta/aorta_classification.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "sqanti_df = pd.read_csv(sqanti_classification, sep='\\t')\n",
    "# Rename and make boolean\n",
    "sqanti_df['category'] = sqanti_df['structural_category']\n",
    "sqanti_df['TSS ratio'] = sqanti_df['ratio_TSS'] > 1.5\n",
    "sqanti_df['CAGE support'] = sqanti_df['within_CAGE_peak']\n",
    "sqanti_df['polyA site'] = sqanti_df['within_polyA_site']\n",
    "sqanti_df['polyA motif'] = sqanti_df['polyA_motif_found']\n",
    "sqanti_df['start both'] = sqanti_df['TSS ratio'] & sqanti_df['CAGE support']\n",
    "sqanti_df['end both'] = sqanti_df['polyA site'] & sqanti_df['polyA motif']\n",
    "\n",
    "sqanti_df['category'] = sqanti_df['category'].replace({\n",
    "    'intergenic': 'Intergenic',\n",
    "    'antisense': 'Antisense',\n",
    "    'genic': 'Genic',\n",
    "    'genic_intron': 'Genic Intron',\n",
    "    'fusion': 'Fusion',\n",
    "    'full-splice_match': 'FSM',\n",
    "    'incomplete-splice_match': 'ISM',\n",
    "    'novel_in_catalog': 'NIC',\n",
    "    'novel_not_in_catalog': 'NNC'\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>isoform</th>\n",
       "      <th>category</th>\n",
       "      <th>subcategory</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000037502.11</td>\n",
       "      <td>FSM</td>\n",
       "      <td>reference_match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000040877.2</td>\n",
       "      <td>FSM</td>\n",
       "      <td>reference_match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000054666.11</td>\n",
       "      <td>FSM</td>\n",
       "      <td>reference_match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000060969.6</td>\n",
       "      <td>FSM</td>\n",
       "      <td>alternative_3end5end</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000072644.7</td>\n",
       "      <td>FSM</td>\n",
       "      <td>reference_match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109140</th>\n",
       "      <td>m54284U_210227_205004/70453536/ccs</td>\n",
       "      <td>NIC</td>\n",
       "      <td>combination_of_known_splicesites</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109141</th>\n",
       "      <td>m54284U_210227_205004/722100/ccs</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109142</th>\n",
       "      <td>m54284U_210227_205004/72286900/ccs</td>\n",
       "      <td>Genic Intron</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109143</th>\n",
       "      <td>m54284U_210227_205004/83820795/ccs</td>\n",
       "      <td>Genic Intron</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109144</th>\n",
       "      <td>m54284U_210227_205004/88410435/ccs</td>\n",
       "      <td>Antisense</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>109145 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   isoform      category  \\\n",
       "0                       ENST00000037502.11           FSM   \n",
       "1                        ENST00000040877.2           FSM   \n",
       "2                       ENST00000054666.11           FSM   \n",
       "3                        ENST00000060969.6           FSM   \n",
       "4                        ENST00000072644.7           FSM   \n",
       "...                                    ...           ...   \n",
       "109140  m54284U_210227_205004/70453536/ccs           NIC   \n",
       "109141    m54284U_210227_205004/722100/ccs    Intergenic   \n",
       "109142  m54284U_210227_205004/72286900/ccs  Genic Intron   \n",
       "109143  m54284U_210227_205004/83820795/ccs  Genic Intron   \n",
       "109144  m54284U_210227_205004/88410435/ccs     Antisense   \n",
       "\n",
       "                             subcategory  \n",
       "0                        reference_match  \n",
       "1                        reference_match  \n",
       "2                        reference_match  \n",
       "3                   alternative_3end5end  \n",
       "4                        reference_match  \n",
       "...                                  ...  \n",
       "109140  combination_of_known_splicesites  \n",
       "109141                         mono-exon  \n",
       "109142                         mono-exon  \n",
       "109143                         mono-exon  \n",
       "109144                         mono-exon  \n",
       "\n",
       "[109145 rows x 3 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sqanti_df = sqanti_df.loc[:,['isoform', 'category', 'subcategory']]\n",
    "sqanti_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "isoseq = Transcriptome.load(isotools_pkl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "isotools_df = isoseq.transcript_table(query=\"\", extra_columns=[\"transcript_id\"])\\\n",
    "# check if transcript_id is unique, otherwise\n",
    "# create isoform column from gene_id + _ + transcript_nr\n",
    "if len(isotools_df['transcript_id'].unique()) == isotools_df.shape[0]:\n",
    "    isotools_df = isotools_df.assign(isoform=isotools_df['transcript_id'])\n",
    "else:\n",
    "    isotools_df = isotools_df.assign(isoform=isotools_df['gene_id'] + '_' + isotools_df['transcript_nr'].astype(str))\n",
    "# rename novelty_class to isotools_category and novelty_subclasses to isotools_subcategory\n",
    "isotools_df = isotools_df.rename(columns={'novelty_class': 'isotools_category', 'novelty_subclasses': 'isotools_subcategory'})\n",
    "isotools_df = isotools_df.loc[:,['gene_id', 'isoform', 'isotools_category', 'isotools_subcategory', 'chr', 'transcript_start', 'transcript_end']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>isoform</th>\n",
       "      <th>isotools_category</th>\n",
       "      <th>isotools_subcategory</th>\n",
       "      <th>chr</th>\n",
       "      <th>transcript_start</th>\n",
       "      <th>transcript_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000121957.15</td>\n",
       "      <td>m54284U_210227_205004/37357274/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108876984</td>\n",
       "      <td>108880418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000121957.15</td>\n",
       "      <td>m54284U_210227_205004/61016242/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108906669</td>\n",
       "      <td>108909981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000121957.15</td>\n",
       "      <td>m54284U_210227_205004/84346647/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108925726</td>\n",
       "      <td>108926898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000121957.15</td>\n",
       "      <td>m54284U_210227_205004/103220783/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108928073</td>\n",
       "      <td>108931970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000121957.15</td>\n",
       "      <td>ENST00000264126.9</td>\n",
       "      <td>FSM</td>\n",
       "      <td>FSM</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108876984</td>\n",
       "      <td>108930382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107486</th>\n",
       "      <td>IT_novel_53673</td>\n",
       "      <td>m54284U_200419_012632/115279072/ccs</td>\n",
       "      <td>NOVEL</td>\n",
       "      <td>antisense</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1385100</td>\n",
       "      <td>1386656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107487</th>\n",
       "      <td>ENSG00000292335.1</td>\n",
       "      <td>m54284U_210227_205004/121636612/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chrY</td>\n",
       "      <td>992801</td>\n",
       "      <td>994365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107488</th>\n",
       "      <td>IT_novel_53657</td>\n",
       "      <td>m54284U_210227_205004/20318378/ccs</td>\n",
       "      <td>NOVEL</td>\n",
       "      <td>intronic</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1405441</td>\n",
       "      <td>1407766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107489</th>\n",
       "      <td>IT_novel_53664</td>\n",
       "      <td>m54284U_200419_012632/12716168/ccs</td>\n",
       "      <td>NOVEL</td>\n",
       "      <td>genic genomic</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1415645</td>\n",
       "      <td>1418144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107490</th>\n",
       "      <td>ENSG00000292342.1</td>\n",
       "      <td>m54284U_200419_012632/106826569/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chrY</td>\n",
       "      <td>319175</td>\n",
       "      <td>321317</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>107491 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_id                              isoform  \\\n",
       "0       ENSG00000121957.15   m54284U_210227_205004/37357274/ccs   \n",
       "1       ENSG00000121957.15   m54284U_210227_205004/61016242/ccs   \n",
       "2       ENSG00000121957.15   m54284U_210227_205004/84346647/ccs   \n",
       "3       ENSG00000121957.15  m54284U_210227_205004/103220783/ccs   \n",
       "4       ENSG00000121957.15                    ENST00000264126.9   \n",
       "...                    ...                                  ...   \n",
       "107486      IT_novel_53673  m54284U_200419_012632/115279072/ccs   \n",
       "107487   ENSG00000292335.1  m54284U_210227_205004/121636612/ccs   \n",
       "107488      IT_novel_53657   m54284U_210227_205004/20318378/ccs   \n",
       "107489      IT_novel_53664   m54284U_200419_012632/12716168/ccs   \n",
       "107490   ENSG00000292342.1  m54284U_200419_012632/106826569/ccs   \n",
       "\n",
       "       isotools_category isotools_subcategory   chr  transcript_start  \\\n",
       "0                    ISM            mono-exon  chr1         108876984   \n",
       "1                    ISM            mono-exon  chr1         108906669   \n",
       "2                    ISM            mono-exon  chr1         108925726   \n",
       "3                    ISM            mono-exon  chr1         108928073   \n",
       "4                    FSM                  FSM  chr1         108876984   \n",
       "...                  ...                  ...   ...               ...   \n",
       "107486             NOVEL            antisense  chrY           1385100   \n",
       "107487               ISM            mono-exon  chrY            992801   \n",
       "107488             NOVEL             intronic  chrY           1405441   \n",
       "107489             NOVEL        genic genomic  chrY           1415645   \n",
       "107490               ISM            mono-exon  chrY            319175   \n",
       "\n",
       "        transcript_end  \n",
       "0            108880418  \n",
       "1            108909981  \n",
       "2            108926898  \n",
       "3            108931970  \n",
       "4            108930382  \n",
       "...                ...  \n",
       "107486         1386656  \n",
       "107487          994365  \n",
       "107488         1407766  \n",
       "107489         1418144  \n",
       "107490          321317  \n",
       "\n",
       "[107491 rows x 7 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isotools_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>isoform</th>\n",
       "      <th>isotools_category</th>\n",
       "      <th>isotools_subcategory</th>\n",
       "      <th>chr</th>\n",
       "      <th>transcript_start</th>\n",
       "      <th>transcript_end</th>\n",
       "      <th>category</th>\n",
       "      <th>subcategory</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000121957.15</td>\n",
       "      <td>m54284U_210227_205004/37357274/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108876984</td>\n",
       "      <td>108880418</td>\n",
       "      <td>NIC</td>\n",
       "      <td>mono-exon_by_intron_retention</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000121957.15</td>\n",
       "      <td>m54284U_210227_205004/61016242/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108906669</td>\n",
       "      <td>108909981</td>\n",
       "      <td>Genic</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000121957.15</td>\n",
       "      <td>m54284U_210227_205004/84346647/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108925726</td>\n",
       "      <td>108926898</td>\n",
       "      <td>Genic</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000121957.15</td>\n",
       "      <td>m54284U_210227_205004/103220783/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108928073</td>\n",
       "      <td>108931970</td>\n",
       "      <td>NIC</td>\n",
       "      <td>mono-exon_by_intron_retention</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000121957.15</td>\n",
       "      <td>ENST00000264126.9</td>\n",
       "      <td>FSM</td>\n",
       "      <td>FSM</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108876984</td>\n",
       "      <td>108930382</td>\n",
       "      <td>FSM</td>\n",
       "      <td>alternative_3end</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107486</th>\n",
       "      <td>IT_novel_53673</td>\n",
       "      <td>m54284U_200419_012632/115279072/ccs</td>\n",
       "      <td>NOVEL</td>\n",
       "      <td>antisense</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1385100</td>\n",
       "      <td>1386656</td>\n",
       "      <td>Antisense</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107487</th>\n",
       "      <td>ENSG00000292335.1</td>\n",
       "      <td>m54284U_210227_205004/121636612/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chrY</td>\n",
       "      <td>992801</td>\n",
       "      <td>994365</td>\n",
       "      <td>Genic</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107488</th>\n",
       "      <td>IT_novel_53657</td>\n",
       "      <td>m54284U_210227_205004/20318378/ccs</td>\n",
       "      <td>NOVEL</td>\n",
       "      <td>intronic</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1405441</td>\n",
       "      <td>1407766</td>\n",
       "      <td>Genic Intron</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107489</th>\n",
       "      <td>IT_novel_53664</td>\n",
       "      <td>m54284U_200419_012632/12716168/ccs</td>\n",
       "      <td>NOVEL</td>\n",
       "      <td>genic genomic</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1415645</td>\n",
       "      <td>1418144</td>\n",
       "      <td>Genic</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107490</th>\n",
       "      <td>ENSG00000292342.1</td>\n",
       "      <td>m54284U_200419_012632/106826569/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chrY</td>\n",
       "      <td>319175</td>\n",
       "      <td>321317</td>\n",
       "      <td>Genic</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>107491 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_id                              isoform  \\\n",
       "0       ENSG00000121957.15   m54284U_210227_205004/37357274/ccs   \n",
       "1       ENSG00000121957.15   m54284U_210227_205004/61016242/ccs   \n",
       "2       ENSG00000121957.15   m54284U_210227_205004/84346647/ccs   \n",
       "3       ENSG00000121957.15  m54284U_210227_205004/103220783/ccs   \n",
       "4       ENSG00000121957.15                    ENST00000264126.9   \n",
       "...                    ...                                  ...   \n",
       "107486      IT_novel_53673  m54284U_200419_012632/115279072/ccs   \n",
       "107487   ENSG00000292335.1  m54284U_210227_205004/121636612/ccs   \n",
       "107488      IT_novel_53657   m54284U_210227_205004/20318378/ccs   \n",
       "107489      IT_novel_53664   m54284U_200419_012632/12716168/ccs   \n",
       "107490   ENSG00000292342.1  m54284U_200419_012632/106826569/ccs   \n",
       "\n",
       "       isotools_category isotools_subcategory   chr  transcript_start  \\\n",
       "0                    ISM            mono-exon  chr1         108876984   \n",
       "1                    ISM            mono-exon  chr1         108906669   \n",
       "2                    ISM            mono-exon  chr1         108925726   \n",
       "3                    ISM            mono-exon  chr1         108928073   \n",
       "4                    FSM                  FSM  chr1         108876984   \n",
       "...                  ...                  ...   ...               ...   \n",
       "107486             NOVEL            antisense  chrY           1385100   \n",
       "107487               ISM            mono-exon  chrY            992801   \n",
       "107488             NOVEL             intronic  chrY           1405441   \n",
       "107489             NOVEL        genic genomic  chrY           1415645   \n",
       "107490               ISM            mono-exon  chrY            319175   \n",
       "\n",
       "        transcript_end      category                    subcategory  \n",
       "0            108880418           NIC  mono-exon_by_intron_retention  \n",
       "1            108909981         Genic                      mono-exon  \n",
       "2            108926898         Genic                      mono-exon  \n",
       "3            108931970           NIC  mono-exon_by_intron_retention  \n",
       "4            108930382           FSM               alternative_3end  \n",
       "...                ...           ...                            ...  \n",
       "107486         1386656     Antisense                      mono-exon  \n",
       "107487          994365         Genic                      mono-exon  \n",
       "107488         1407766  Genic Intron                      mono-exon  \n",
       "107489         1418144         Genic                      mono-exon  \n",
       "107490          321317         Genic                      mono-exon  \n",
       "\n",
       "[107491 rows x 9 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# join dfs\n",
    "df = pd.merge(isotools_df, sqanti_df, on='isoform', how='inner')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>isoform</th>\n",
       "      <th>isotools_category</th>\n",
       "      <th>isotools_subcategory</th>\n",
       "      <th>chr</th>\n",
       "      <th>transcript_start</th>\n",
       "      <th>transcript_end</th>\n",
       "      <th>category</th>\n",
       "      <th>subcategory</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000121957.15</td>\n",
       "      <td>m54284U_210227_205004/37357274/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108876984</td>\n",
       "      <td>108880418</td>\n",
       "      <td>NIC</td>\n",
       "      <td>mono-exon_by_intron_retention</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000121957.15</td>\n",
       "      <td>m54284U_210227_205004/61016242/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108906669</td>\n",
       "      <td>108909981</td>\n",
       "      <td>Genic</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000121957.15</td>\n",
       "      <td>m54284U_210227_205004/84346647/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108925726</td>\n",
       "      <td>108926898</td>\n",
       "      <td>Genic</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000121957.15</td>\n",
       "      <td>m54284U_210227_205004/103220783/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108928073</td>\n",
       "      <td>108931970</td>\n",
       "      <td>NIC</td>\n",
       "      <td>mono-exon_by_intron_retention</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000163431.14</td>\n",
       "      <td>m54284U_200419_012632/53545803/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chr1</td>\n",
       "      <td>201896455</td>\n",
       "      <td>201898889</td>\n",
       "      <td>Genic</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107424</th>\n",
       "      <td>ENSG00000157514.18</td>\n",
       "      <td>m54284U_200419_012632/25363560/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chrX</td>\n",
       "      <td>107713223</td>\n",
       "      <td>107716269</td>\n",
       "      <td>NIC</td>\n",
       "      <td>mono-exon_by_intron_retention</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107444</th>\n",
       "      <td>ENSG00000292341.1</td>\n",
       "      <td>m54284U_210227_205004/123078285/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1398288</td>\n",
       "      <td>1399412</td>\n",
       "      <td>NIC</td>\n",
       "      <td>mono-exon_by_intron_retention</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107455</th>\n",
       "      <td>ENSG00000292357.1</td>\n",
       "      <td>m54284U_210227_205004/33751947/ccs</td>\n",
       "      <td>NIC</td>\n",
       "      <td>intron retention</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1285651</td>\n",
       "      <td>1307499</td>\n",
       "      <td>ISM</td>\n",
       "      <td>internal_fragment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107487</th>\n",
       "      <td>ENSG00000292335.1</td>\n",
       "      <td>m54284U_210227_205004/121636612/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chrY</td>\n",
       "      <td>992801</td>\n",
       "      <td>994365</td>\n",
       "      <td>Genic</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107490</th>\n",
       "      <td>ENSG00000292342.1</td>\n",
       "      <td>m54284U_200419_012632/106826569/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chrY</td>\n",
       "      <td>319175</td>\n",
       "      <td>321317</td>\n",
       "      <td>Genic</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7665 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_id                              isoform  \\\n",
       "0       ENSG00000121957.15   m54284U_210227_205004/37357274/ccs   \n",
       "1       ENSG00000121957.15   m54284U_210227_205004/61016242/ccs   \n",
       "2       ENSG00000121957.15   m54284U_210227_205004/84346647/ccs   \n",
       "3       ENSG00000121957.15  m54284U_210227_205004/103220783/ccs   \n",
       "7       ENSG00000163431.14   m54284U_200419_012632/53545803/ccs   \n",
       "...                    ...                                  ...   \n",
       "107424  ENSG00000157514.18   m54284U_200419_012632/25363560/ccs   \n",
       "107444   ENSG00000292341.1  m54284U_210227_205004/123078285/ccs   \n",
       "107455   ENSG00000292357.1   m54284U_210227_205004/33751947/ccs   \n",
       "107487   ENSG00000292335.1  m54284U_210227_205004/121636612/ccs   \n",
       "107490   ENSG00000292342.1  m54284U_200419_012632/106826569/ccs   \n",
       "\n",
       "       isotools_category isotools_subcategory   chr  transcript_start  \\\n",
       "0                    ISM            mono-exon  chr1         108876984   \n",
       "1                    ISM            mono-exon  chr1         108906669   \n",
       "2                    ISM            mono-exon  chr1         108925726   \n",
       "3                    ISM            mono-exon  chr1         108928073   \n",
       "7                    ISM            mono-exon  chr1         201896455   \n",
       "...                  ...                  ...   ...               ...   \n",
       "107424               ISM            mono-exon  chrX         107713223   \n",
       "107444               ISM            mono-exon  chrY           1398288   \n",
       "107455               NIC     intron retention  chrY           1285651   \n",
       "107487               ISM            mono-exon  chrY            992801   \n",
       "107490               ISM            mono-exon  chrY            319175   \n",
       "\n",
       "        transcript_end category                    subcategory  \n",
       "0            108880418      NIC  mono-exon_by_intron_retention  \n",
       "1            108909981    Genic                      mono-exon  \n",
       "2            108926898    Genic                      mono-exon  \n",
       "3            108931970      NIC  mono-exon_by_intron_retention  \n",
       "7            201898889    Genic                      mono-exon  \n",
       "...                ...      ...                            ...  \n",
       "107424       107716269      NIC  mono-exon_by_intron_retention  \n",
       "107444         1399412      NIC  mono-exon_by_intron_retention  \n",
       "107455         1307499      ISM              internal_fragment  \n",
       "107487          994365    Genic                      mono-exon  \n",
       "107490          321317    Genic                      mono-exon  \n",
       "\n",
       "[7665 rows x 9 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mismatches = df.loc[(df['category'] != df['isotools_category']) & (df['isotools_category'] != 'NOVEL')]\n",
    "mismatches"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Monoexons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>isoform</th>\n",
       "      <th>isotools_category</th>\n",
       "      <th>isotools_subcategory</th>\n",
       "      <th>chr</th>\n",
       "      <th>transcript_start</th>\n",
       "      <th>transcript_end</th>\n",
       "      <th>category</th>\n",
       "      <th>subcategory</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000121957.15</td>\n",
       "      <td>m54284U_210227_205004/61016242/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108906669</td>\n",
       "      <td>108909981</td>\n",
       "      <td>Genic</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000121957.15</td>\n",
       "      <td>m54284U_210227_205004/84346647/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108925726</td>\n",
       "      <td>108926898</td>\n",
       "      <td>Genic</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000163431.14</td>\n",
       "      <td>m54284U_200419_012632/53545803/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chr1</td>\n",
       "      <td>201896455</td>\n",
       "      <td>201898889</td>\n",
       "      <td>Genic</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>ENSG00000116685.17</td>\n",
       "      <td>m54284U_210227_205004/122096460/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11919590</td>\n",
       "      <td>11922106</td>\n",
       "      <td>Genic</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>ENSG00000251484.3</td>\n",
       "      <td>m54284U_200419_012632/112657704/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chr1</td>\n",
       "      <td>109103706</td>\n",
       "      <td>109105540</td>\n",
       "      <td>Genic</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107148</th>\n",
       "      <td>ENSG00000236017.8</td>\n",
       "      <td>m54284U_210227_205004/64290982/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chrX</td>\n",
       "      <td>1412900</td>\n",
       "      <td>1414741</td>\n",
       "      <td>Genic</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107272</th>\n",
       "      <td>ENSG00000233661.1</td>\n",
       "      <td>m54284U_210227_205004/4589284/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chrX</td>\n",
       "      <td>63351095</td>\n",
       "      <td>63356012</td>\n",
       "      <td>Genic</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107360</th>\n",
       "      <td>ENSG00000268738.3</td>\n",
       "      <td>m54284U_200419_012632/93651732/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chrX</td>\n",
       "      <td>149592043</td>\n",
       "      <td>149593938</td>\n",
       "      <td>Genic</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107487</th>\n",
       "      <td>ENSG00000292335.1</td>\n",
       "      <td>m54284U_210227_205004/121636612/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chrY</td>\n",
       "      <td>992801</td>\n",
       "      <td>994365</td>\n",
       "      <td>Genic</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107490</th>\n",
       "      <td>ENSG00000292342.1</td>\n",
       "      <td>m54284U_200419_012632/106826569/ccs</td>\n",
       "      <td>ISM</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>chrY</td>\n",
       "      <td>319175</td>\n",
       "      <td>321317</td>\n",
       "      <td>Genic</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3389 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_id                              isoform  \\\n",
       "1       ENSG00000121957.15   m54284U_210227_205004/61016242/ccs   \n",
       "2       ENSG00000121957.15   m54284U_210227_205004/84346647/ccs   \n",
       "7       ENSG00000163431.14   m54284U_200419_012632/53545803/ccs   \n",
       "34      ENSG00000116685.17  m54284U_210227_205004/122096460/ccs   \n",
       "39       ENSG00000251484.3  m54284U_200419_012632/112657704/ccs   \n",
       "...                    ...                                  ...   \n",
       "107148   ENSG00000236017.8   m54284U_210227_205004/64290982/ccs   \n",
       "107272   ENSG00000233661.1    m54284U_210227_205004/4589284/ccs   \n",
       "107360   ENSG00000268738.3   m54284U_200419_012632/93651732/ccs   \n",
       "107487   ENSG00000292335.1  m54284U_210227_205004/121636612/ccs   \n",
       "107490   ENSG00000292342.1  m54284U_200419_012632/106826569/ccs   \n",
       "\n",
       "       isotools_category isotools_subcategory   chr  transcript_start  \\\n",
       "1                    ISM            mono-exon  chr1         108906669   \n",
       "2                    ISM            mono-exon  chr1         108925726   \n",
       "7                    ISM            mono-exon  chr1         201896455   \n",
       "34                   ISM            mono-exon  chr1          11919590   \n",
       "39                   ISM            mono-exon  chr1         109103706   \n",
       "...                  ...                  ...   ...               ...   \n",
       "107148               ISM            mono-exon  chrX           1412900   \n",
       "107272               ISM            mono-exon  chrX          63351095   \n",
       "107360               ISM            mono-exon  chrX         149592043   \n",
       "107487               ISM            mono-exon  chrY            992801   \n",
       "107490               ISM            mono-exon  chrY            319175   \n",
       "\n",
       "        transcript_end category subcategory  \n",
       "1            108909981    Genic   mono-exon  \n",
       "2            108926898    Genic   mono-exon  \n",
       "7            201898889    Genic   mono-exon  \n",
       "34            11922106    Genic   mono-exon  \n",
       "39           109105540    Genic   mono-exon  \n",
       "...                ...      ...         ...  \n",
       "107148         1414741    Genic   mono-exon  \n",
       "107272        63356012    Genic   mono-exon  \n",
       "107360       149593938    Genic   mono-exon  \n",
       "107487          994365    Genic   mono-exon  \n",
       "107490          321317    Genic   mono-exon  \n",
       "\n",
       "[3389 rows x 9 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "monoexon_mm = mismatches.loc[mismatches['subcategory'] == 'mono-exon']\n",
    "monoexon_mm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[13], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# First transcript from the list\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m \u001b[43misoseq\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mENSG00000121957.15\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtranscripts\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m13\u001b[39;49m\u001b[43m]\u001b[49m\n",
      "\u001b[0;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "# First transcript from the list\n",
    "isoseq['ENSG00000121957.15'].transcripts[13]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "isodev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
