{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('/home/lankenau/isotools/src')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:This is isotools version 0.3.5rc11\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import logging\n",
    "from isotools import Gene\n",
    "from isotools import Transcriptome\n",
    "from isotools import __version__ as isotools_version\n",
    "\n",
    "# set up logging\n",
    "logging.basicConfig(format=\"%(levelname)s:%(message)s\", level=logging.INFO)\n",
    "logger = logging.getLogger(\"isotools\")\n",
    "logger.info(f\"This is isotools version {isotools_version}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:loading transcriptome from /project/hfa_work/ENCODE/code/snakemake-pipeline/results/isotools_v2/pkl/brain.pkl\n"
     ]
    }
   ],
   "source": [
    "\n",
    "basepath = '/project/hfa_work/ENCODE/code/snakemake-pipeline/'\n",
    "version = 2\n",
    "tissue = 'brain'\n",
    "path = os.path.join(basepath, f\"results/isotools_v{version}/pkl/{tissue}.pkl\")\n",
    "isoseq = Transcriptome.load(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 55811/55811 [00:19<00:00, 2924.26it/s]\n"
     ]
    }
   ],
   "source": [
    "classification = os.path.join(basepath, 'results/sqanti/isotools_v{version}/qc/{tissue}/brain_classification.txt')\n",
    "isoseq.import_sqanti_classification(classification.format(version=version, tissue=tissue))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "isoseq.add_filter(tag='CAGE_SUPPORT', expression='sqanti_classification is not None and sqanti_classification[\"within_CAGE_peak\"]', context='transcript', update=True)\n",
    "isoseq.add_filter(tag='TSS_RATIO', expression='sqanti_classification is not None and sqanti_classification[\"ratio_TSS\"] > 1.5', context='transcript', update=True)\n",
    "isoseq.add_filter(tag='POLYA_SITE', expression='sqanti_classification is not None and sqanti_classification[\"within_polyA_site\"]', context='transcript', update=True)\n",
    "isoseq.add_filter(tag='POLYA_MOTIF', expression='sqanti_classification is not None and sqanti_classification[\"polyA_motif_found\"]', context='transcript', update=True)\n",
    "\n",
    "isoseq.add_filter(tag='SQANTI_SUPPORT', expression='CAGE_SUPPORT and TSS_RATIO and POLYA_SITE and POLYA_MOTIF', context='transcript', update=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(<isotools.gene.Gene object at 0x7f82835be870>,\n",
       " 0,\n",
       " {'exons': [[np.int64(108876985), 108877228],\n",
       "   [108885506, 108885578],\n",
       "   [108896863, 108897085],\n",
       "   [108897491, 108897627],\n",
       "   [108897958, 108898101],\n",
       "   [108898641, 108898765],\n",
       "   [108898878, 108898994],\n",
       "   [108901789, 108901945],\n",
       "   [108903125, 108903234],\n",
       "   [108904124, 108904254],\n",
       "   [108914337, 108914408],\n",
       "   [108918612, 108918789],\n",
       "   [108922416, 108922576],\n",
       "   [108923999, 108924214],\n",
       "   [108929700, np.int64(108930395)]],\n",
       "  'strand': '+',\n",
       "  'coverage': {'ENCSR700EBI': 10, 'ENCSR425HFS': 1, 'ENCSR463IDK': 10},\n",
       "  'TSS': {'ENCSR700EBI': {108876981: 1,\n",
       "    108876984: 1,\n",
       "    108876985: 2,\n",
       "    108876986: 1,\n",
       "    108876988: 2,\n",
       "    108876990: 1,\n",
       "    108877106: 1,\n",
       "    108877146: 1},\n",
       "   'ENCSR425HFS': {108876984: 1},\n",
       "   'ENCSR463IDK': {108876984: 5,\n",
       "    108876985: 1,\n",
       "    108876987: 2,\n",
       "    108877098: 1,\n",
       "    108877168: 1}},\n",
       "  'PAS': {'ENCSR700EBI': {108930396: 7,\n",
       "    108931627: 1,\n",
       "    108930404: 1,\n",
       "    108930391: 1},\n",
       "   'ENCSR425HFS': {108930395: 1},\n",
       "   'ENCSR463IDK': {108930396: 5, 108930395: 3, 108930421: 2}},\n",
       "  'annotation': (2, {'novel combination': []}),\n",
       "  'TSS_unified': {'ENCSR700EBI': {np.int64(108876985): 10},\n",
       "   'ENCSR425HFS': {np.int64(108876985): 1},\n",
       "   'ENCSR463IDK': {np.int64(108876985): 10}},\n",
       "  'PAS_unified': {'ENCSR700EBI': {np.int64(108930395): 10},\n",
       "   'ENCSR425HFS': {np.int64(108930395): 1},\n",
       "   'ENCSR463IDK': {np.int64(108930395): 10}},\n",
       "  'fragments': [(9, 0, 1)],\n",
       "  'direct_repeat_len': [4, 4, 5, 4, 5, 3, 4, 4, 5, 5, 4, 2, 3, 4],\n",
       "  'downstream_A_content': 0.26666666666666666,\n",
       "  'ORF': (np.int64(108885522),\n",
       "   np.int64(108929940),\n",
       "   {\"5'UTR\": np.int64(259),\n",
       "    'CDS': np.int64(2055),\n",
       "    \"3'UTR\": np.int64(455),\n",
       "    'start_codon': 'ATG',\n",
       "    'stop_codon': 'TAG',\n",
       "    'NMD': False,\n",
       "    'uORFs': 0,\n",
       "    'ref_ids': [18],\n",
       "    'kozak': np.float64(-1.271810457563484),\n",
       "    'fickett': 0.9859}),\n",
       "  'sqanti_classification': {'dist_to_CAGE_peak': 0.0,\n",
       "   'within_CAGE_peak': True,\n",
       "   'dist_to_polyA_site': 9.0,\n",
       "   'within_polyA_site': True,\n",
       "   'polyA_motif': 'TATAAA',\n",
       "   'polyA_dist': -11.0,\n",
       "   'polyA_motif_found': True,\n",
       "   'ratio_TSS': 2.8798374204820534}})"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "next(isoseq.iter_transcripts(query='SQANTI_SUPPORT'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11661"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count = 0\n",
    "for transcript in isoseq.iter_transcripts(query='SQANTI_SUPPORT'):\n",
    "    count += 1\n",
    "count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>strand</th>\n",
       "      <th>eventA_type</th>\n",
       "      <th>eventB_type</th>\n",
       "      <th>eventA_start</th>\n",
       "      <th>eventA_end</th>\n",
       "      <th>eventB_start</th>\n",
       "      <th>eventB_end</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>...</th>\n",
       "      <th>dcPSI_AB</th>\n",
       "      <th>dcPSI_BA</th>\n",
       "      <th>priA_priB</th>\n",
       "      <th>priA_altB</th>\n",
       "      <th>altA_priB</th>\n",
       "      <th>altA_altB</th>\n",
       "      <th>priA_priB_transcript_ids</th>\n",
       "      <th>priA_altB_transcript_ids</th>\n",
       "      <th>altA_priB_transcript_ids</th>\n",
       "      <th>altA_altB_transcript_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000142875.21</td>\n",
       "      <td>PRKACB</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS</td>\n",
       "      <td>PAS</td>\n",
       "      <td>84078084</td>\n",
       "      <td>84078371</td>\n",
       "      <td>84204481</td>\n",
       "      <td>84205294</td>\n",
       "      <td>8.387972e-15</td>\n",
       "      <td>...</td>\n",
       "      <td>0.458757</td>\n",
       "      <td>0.491525</td>\n",
       "      <td>29</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>[4, 3, 2]</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000142875.21</td>\n",
       "      <td>PRKACB</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS</td>\n",
       "      <td>PAS</td>\n",
       "      <td>84078084</td>\n",
       "      <td>84078371</td>\n",
       "      <td>84235179</td>\n",
       "      <td>84238491</td>\n",
       "      <td>8.387972e-15</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.458757</td>\n",
       "      <td>-0.443958</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>29</td>\n",
       "      <td>2</td>\n",
       "      <td>[]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[4, 3, 2]</td>\n",
       "      <td>[0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000142875.21</td>\n",
       "      <td>PRKACB</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS</td>\n",
       "      <td>PAS</td>\n",
       "      <td>84164341</td>\n",
       "      <td>84164422</td>\n",
       "      <td>84204481</td>\n",
       "      <td>84205294</td>\n",
       "      <td>5.356978e-02</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.474576</td>\n",
       "      <td>-0.084746</td>\n",
       "      <td>26</td>\n",
       "      <td>5</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>[4, 0]</td>\n",
       "      <td>[3, 2]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000142875.21</td>\n",
       "      <td>PRKACB</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS</td>\n",
       "      <td>PAS</td>\n",
       "      <td>84164341</td>\n",
       "      <td>84164422</td>\n",
       "      <td>84235179</td>\n",
       "      <td>84238491</td>\n",
       "      <td>5.356978e-02</td>\n",
       "      <td>...</td>\n",
       "      <td>0.474576</td>\n",
       "      <td>0.076545</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>26</td>\n",
       "      <td>5</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[4, 0]</td>\n",
       "      <td>[3, 2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000142875.21</td>\n",
       "      <td>PRKACB</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS</td>\n",
       "      <td>PAS</td>\n",
       "      <td>84164694</td>\n",
       "      <td>84165013</td>\n",
       "      <td>84204481</td>\n",
       "      <td>84205294</td>\n",
       "      <td>1.225092e-10</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.474576</td>\n",
       "      <td>-0.406780</td>\n",
       "      <td>7</td>\n",
       "      <td>24</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>[3, 0, 2]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>ENSG00000172465.14</td>\n",
       "      <td>TCEAL1</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS</td>\n",
       "      <td>PAS</td>\n",
       "      <td>103629476</td>\n",
       "      <td>103629573</td>\n",
       "      <td>103629884</td>\n",
       "      <td>103630946</td>\n",
       "      <td>3.526644e-06</td>\n",
       "      <td>...</td>\n",
       "      <td>0.607843</td>\n",
       "      <td>0.334314</td>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>11</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[2, 0, 4]</td>\n",
       "      <td>[3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>ENSG00000133142.19</td>\n",
       "      <td>TCEAL4</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS</td>\n",
       "      <td>PAS</td>\n",
       "      <td>103585517</td>\n",
       "      <td>103585624</td>\n",
       "      <td>103586648</td>\n",
       "      <td>103587725</td>\n",
       "      <td>1.274413e-63</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.107557</td>\n",
       "      <td>-0.497652</td>\n",
       "      <td>8</td>\n",
       "      <td>570</td>\n",
       "      <td>84</td>\n",
       "      <td>50</td>\n",
       "      <td>[6]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[2, 5]</td>\n",
       "      <td>[3, 0, 4]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251</th>\n",
       "      <td>ENSG00000133142.19</td>\n",
       "      <td>TCEAL4</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS</td>\n",
       "      <td>PAS</td>\n",
       "      <td>103585517</td>\n",
       "      <td>103585624</td>\n",
       "      <td>103586653</td>\n",
       "      <td>103587725</td>\n",
       "      <td>1.274413e-63</td>\n",
       "      <td>...</td>\n",
       "      <td>0.107557</td>\n",
       "      <td>0.115373</td>\n",
       "      <td>84</td>\n",
       "      <td>50</td>\n",
       "      <td>8</td>\n",
       "      <td>570</td>\n",
       "      <td>[2, 5]</td>\n",
       "      <td>[3, 0, 4]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>[1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>ENSG00000133142.19</td>\n",
       "      <td>TCEAL4</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS</td>\n",
       "      <td>PAS</td>\n",
       "      <td>103585517</td>\n",
       "      <td>103585683</td>\n",
       "      <td>103586648</td>\n",
       "      <td>103587725</td>\n",
       "      <td>1.742996e-59</td>\n",
       "      <td>...</td>\n",
       "      <td>0.720889</td>\n",
       "      <td>0.473419</td>\n",
       "      <td>570</td>\n",
       "      <td>8</td>\n",
       "      <td>54</td>\n",
       "      <td>80</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>[3, 5, 0, 4]</td>\n",
       "      <td>[2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253</th>\n",
       "      <td>ENSG00000133142.19</td>\n",
       "      <td>TCEAL4</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS</td>\n",
       "      <td>PAS</td>\n",
       "      <td>103585517</td>\n",
       "      <td>103585683</td>\n",
       "      <td>103586653</td>\n",
       "      <td>103587725</td>\n",
       "      <td>1.742996e-59</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.720889</td>\n",
       "      <td>-0.109755</td>\n",
       "      <td>54</td>\n",
       "      <td>80</td>\n",
       "      <td>570</td>\n",
       "      <td>8</td>\n",
       "      <td>[3, 5, 0, 4]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[6]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>254 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_id gene_name strand eventA_type eventB_type  \\\n",
       "0    ENSG00000142875.21    PRKACB      +         TSS         PAS   \n",
       "1    ENSG00000142875.21    PRKACB      +         TSS         PAS   \n",
       "2    ENSG00000142875.21    PRKACB      +         TSS         PAS   \n",
       "3    ENSG00000142875.21    PRKACB      +         TSS         PAS   \n",
       "4    ENSG00000142875.21    PRKACB      +         TSS         PAS   \n",
       "..                  ...       ...    ...         ...         ...   \n",
       "249  ENSG00000172465.14    TCEAL1      +         TSS         PAS   \n",
       "250  ENSG00000133142.19    TCEAL4      +         TSS         PAS   \n",
       "251  ENSG00000133142.19    TCEAL4      +         TSS         PAS   \n",
       "252  ENSG00000133142.19    TCEAL4      +         TSS         PAS   \n",
       "253  ENSG00000133142.19    TCEAL4      +         TSS         PAS   \n",
       "\n",
       "     eventA_start  eventA_end  eventB_start  eventB_end        pvalue  ...  \\\n",
       "0        84078084    84078371      84204481    84205294  8.387972e-15  ...   \n",
       "1        84078084    84078371      84235179    84238491  8.387972e-15  ...   \n",
       "2        84164341    84164422      84204481    84205294  5.356978e-02  ...   \n",
       "3        84164341    84164422      84235179    84238491  5.356978e-02  ...   \n",
       "4        84164694    84165013      84204481    84205294  1.225092e-10  ...   \n",
       "..            ...         ...           ...         ...           ...  ...   \n",
       "249     103629476   103629573     103629884   103630946  3.526644e-06  ...   \n",
       "250     103585517   103585624     103586648   103587725  1.274413e-63  ...   \n",
       "251     103585517   103585624     103586653   103587725  1.274413e-63  ...   \n",
       "252     103585517   103585683     103586648   103587725  1.742996e-59  ...   \n",
       "253     103585517   103585683     103586653   103587725  1.742996e-59  ...   \n",
       "\n",
       "     dcPSI_AB  dcPSI_BA  priA_priB  priA_altB  altA_priB  altA_altB  \\\n",
       "0    0.458757  0.491525         29          2          0         28   \n",
       "1   -0.458757 -0.443958          0         28         29          2   \n",
       "2   -0.474576 -0.084746         26          5         28          0   \n",
       "3    0.474576  0.076545         28          0         26          5   \n",
       "4   -0.474576 -0.406780          7         24         28          0   \n",
       "..        ...       ...        ...        ...        ...        ...   \n",
       "249  0.607843  0.334314         31          0          9         11   \n",
       "250 -0.107557 -0.497652          8        570         84         50   \n",
       "251  0.107557  0.115373         84         50          8        570   \n",
       "252  0.720889  0.473419        570          8         54         80   \n",
       "253 -0.720889 -0.109755         54         80        570          8   \n",
       "\n",
       "     priA_priB_transcript_ids  priA_altB_transcript_ids  \\\n",
       "0                   [4, 3, 2]                       [0]   \n",
       "1                          []                       [1]   \n",
       "2                      [4, 0]                    [3, 2]   \n",
       "3                         [1]                        []   \n",
       "4                   [3, 0, 2]                       [4]   \n",
       "..                        ...                       ...   \n",
       "249                       [1]                        []   \n",
       "250                       [6]                       [1]   \n",
       "251                    [2, 5]                 [3, 0, 4]   \n",
       "252                       [1]                       [6]   \n",
       "253              [3, 5, 0, 4]                       [2]   \n",
       "\n",
       "     altA_priB_transcript_ids altA_altB_transcript_ids  \n",
       "0                          []                      [1]  \n",
       "1                   [4, 3, 2]                      [0]  \n",
       "2                         [1]                       []  \n",
       "3                      [4, 0]                   [3, 2]  \n",
       "4                         [1]                       []  \n",
       "..                        ...                      ...  \n",
       "249                 [2, 0, 4]                      [3]  \n",
       "250                    [2, 5]                [3, 0, 4]  \n",
       "251                       [6]                      [1]  \n",
       "252              [3, 5, 0, 4]                      [2]  \n",
       "253                       [1]                      [6]  \n",
       "\n",
       "[254 rows x 23 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coordination_results = isoseq.coordination_test(\n",
    "    event_type=[\"TSS\", \"PAS\"],\n",
    "    min_total=50,\n",
    "    min_alt_fraction=.05,\n",
    "    transcript_filter=\"SQANTI_SUPPORT\",\n",
    ")\n",
    "coordination_results"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "isodev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
